{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "orig_nbformat": 2,
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "metadata": {
      "interpreter": {
        "hash": "26c5dd174060f598a853a789f33344615772874efd96031945d3e0c457607b01"
      }
    },
    "colab": {
      "name": "project.ipynb",
      "provenance": [],
      "collapsed_sections": [
        "kIvTaGMoRVSC",
        "3WVICSt5mLVC",
        "6MTGgcBhmLVH",
        "JVZ4nnpsmLVM"
      ]
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bkF_GZSnmLU4"
      },
      "source": [
        "# Predicting phylum level taxonomic idendities of organisms based on codon usage levels\n",
        "\n",
        "---"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "kIvTaGMoRVSC"
      },
      "source": [
        "## Install some required packages used in this entire notebook\n",
        "\n",
        "---\n",
        "The following python packages are used in this notebook. Run the below cell to install them if you havent done so.\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "EAuWnp_TmLVA",
        "outputId": "13c3a59e-d790-4433-acb2-c52f5bed19e8"
      },
      "source": [
        "!pip install wget requests selenium biopython pandas scikit-learn matplotlib"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: wget in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (3.2)\n",
            "Requirement already satisfied: requests in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (2.25.1)\n",
            "Requirement already satisfied: selenium in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (3.141.0)\n",
            "Requirement already satisfied: biopython in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (1.78)\n",
            "Requirement already satisfied: pandas in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (1.2.3)\n",
            "Requirement already satisfied: scikit-learn in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (0.24.1)\n",
            "Requirement already satisfied: matplotlib in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (3.4.1)\n",
            "Requirement already satisfied: numpy in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (from biopython) (1.20.2)\n",
            "Requirement already satisfied: python-dateutil>=2.7 in c:\\users\\yz\\appdata\\roaming\\python\\python39\\site-packages (from matplotlib) (2.8.1)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (from matplotlib) (1.3.1)\n",
            "Requirement already satisfied: pillow>=6.2.0 in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (from matplotlib) (8.2.0)\n",
            "Requirement already satisfied: cycler>=0.10 in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (from matplotlib) (0.10.0)\n",
            "Requirement already satisfied: pyparsing>=2.2.1 in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (from matplotlib) (2.4.7)\n",
            "Requirement already satisfied: six in c:\\users\\yz\\appdata\\roaming\\python\\python39\\site-packages (from cycler>=0.10->matplotlib) (1.15.0)\n",
            "Requirement already satisfied: pytz>=2017.3 in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (from pandas) (2021.1)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (from requests) (2020.12.5)\n",
            "Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (from requests) (1.26.4)\n",
            "Requirement already satisfied: idna<3,>=2.5 in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (from requests) (2.10)\n",
            "Requirement already satisfied: chardet<5,>=3.0.2 in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (from requests) (4.0.0)\n",
            "Requirement already satisfied: joblib>=0.11 in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (from scikit-learn) (1.0.1)\n",
            "Requirement already satisfied: scipy>=0.19.1 in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (from scikit-learn) (1.6.2)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in c:\\users\\yz\\miniconda3\\envs\\project\\lib\\site-packages (from scikit-learn) (2.1.0)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "3WVICSt5mLVC"
      },
      "source": [
        "## Data download process\n",
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "The data is available from https://hive.biochemistry.gwu.edu/dna.cgi?cmd=tissue_codon_usage&id=586358&mode=cocoputs. The Refseq_species.tsv contains all the codon frequencies of many species from NCBI. Firstly I tried to download via the download link with wget."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "9aIo4vlbmLVD",
        "outputId": "38ea8dbc-3dac-4d40-fcdc-e9e4d37ef5b1"
      },
      "source": [
        "import os\n",
        "import wget\n",
        "print(myfile)\n",
        "data_folder = \"./data\"\n",
        "filename = \"codon_usage.tsv\"\n",
        "if not os.path.exists(data_folder):\n",
        "    os.makedirs(data_folder)\n",
        "filepath = os.path.join(data_folder,filename)\n",
        "url = \"https://hive.biochemistry.gwu.edu/dna.cgi?cmd=objFile&ids=586358&filename=Refseq_species.tsv&raw=1\"\n",
        "wget.download(url, filepath)\n",
        "with open(\"data/codon_usage.tsv\", \"r\") as f:\n",
        "    print(f.readlines())"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "<Response [200]>\n",
            "[]\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6dX-Qde7mLVD"
      },
      "source": [
        "However, the above approach did not work and returned an empty file. After examining the download process with the developers tool in Google Chrome, I realise it requires some headers for the GET request. I tried the below copying the headers in but it still did not work."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "wlxdvrX1mLVE",
        "outputId": "aea1c3de-7dec-4d6b-d9b0-b14d6dbc3ca5"
      },
      "source": [
        "import requests\n",
        "headers = {\n",
        "\"Host\": \"hive.biochemistry.gwu.edu\",\n",
        "\"Connection\": \"keep-alive\",\n",
        "\"sec-ch-ua\": '\"Google Chrome\";v=\"89\", \"Chromium\";v=\"89\", \";Not A Brand\";v=\"99\"',\n",
        "\"sec-ch-ua-mobile\": \"?0\",\n",
        "\"Upgrade-Insecure-Requests\": \"1\",\n",
        "\"User-Agent\": \"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.114 Safari/537.36\",\n",
        "\"Accept\": \"text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9\",\n",
        "\"Sec-Fetch-Site\": \"same-origin\",\n",
        "\"Sec-Fetch-Mode\": \"navigate\",\n",
        "\"Sec-Fetch-User\": \"?1\",\n",
        "\"Sec-Fetch-Dest\": \"document\",\n",
        "\"Referer\": \"https://hive.biochemistry.gwu.edu/dna.cgi?cmd=tissue_codon_usage&id=586358&mode=cocoputs\",\n",
        "\"Accept-Encoding\": \"gzip, deflate, br\",\n",
        "\"Accept-Language\": \"en-US,en;q=0.9\",\n",
        "\n",
        "\"Range\": \"bytes=783482-783482\",\n",
        "\"If-Range\": \"1578517013:86024007\"}\n",
        "myfile = requests.get(\"https://hive.biochemistry.gwu.edu/dna.cgi?cmd=objFile&ids=586358&filename=Refseq_species.tsv&raw=1\", headers=headers)\n",
        "print(myfile.content)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "b''\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "LlrXM2DlmLVF"
      },
      "source": [
        "Once again, it is empty. Hence, I resorted to using Selenium and navigating the browser. First, I downloaded the Chrome driver. Then, I used the browser developer tools to examine the webpage and copy the XPATH of the elements that have to be clicked. On normal navigation, it requires 2 clicks to initate the download of the file. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "OYvFu7qjmLVF",
        "outputId": "e2ae7535-817c-40e0-a098-5ed3a7c73c7d"
      },
      "source": [
        "from zipfile import ZipFile\n",
        "from selenium import webdriver\n",
        "from selenium.webdriver.common.keys import Keys\n",
        "from selenium.webdriver.support import expected_conditions as EC\n",
        "from selenium.webdriver.support.ui import WebDriverWait\n",
        "from selenium.webdriver.common.by import By\n",
        "from selenium.common.exceptions import TimeoutException\n",
        "\n",
        "#downloading and extracting the chrome drive\n",
        "wget.download(\"https://chromedriver.storage.googleapis.com/89.0.4389.23/chromedriver_win32.zip\")\n",
        "file_name = \"chromedriver_win32.zip\"\n",
        "with ZipFile(file_name, 'r') as zip:\n",
        "    zip.extractall()\n",
        "\n",
        "driver = webdriver.Chrome(executable_path=r'./chromedriver.exe')\n",
        "driver.get(\"https://hive.biochemistry.gwu.edu/dna.cgi?cmd=tissue_codon_usage&id=586358&mode=cocoputs\")\n",
        "timeout = 60\n",
        "try:\n",
        "    element_present = EC.presence_of_element_located((By.XPATH, '//*[@id=\"results\"]/li[3]/a'))\n",
        "    WebDriverWait(driver, timeout).until(element_present)\n",
        "    download_tab = driver.find_element_by_xpath(\"/html/body/form[2]/div/div/div[4]/div[1]/div[1]/div[2]/ul/li[3]/ul/li[3]/a\")\n",
        "    download_tab.click()\n",
        "    print(\"Clicked the first button!\")\n",
        "except TimeoutException:\n",
        "    print (\"Timed out waiting for page to load\")\n",
        "\n",
        "try:\n",
        "    element_present = EC.presence_of_element_located((By.XPATH, '/html/body/form[2]/div/div/div[7]/div[1]/div[2]/div/div/div[2]/table/tbody/tr[18]/td[2]'))\n",
        "    WebDriverWait(driver, timeout).until(element_present)\n",
        "    download = driver.find_element_by_xpath('/html/body/form[2]/div/div/div[7]/div[1]/div[2]/div/div/div[2]/table/tbody/tr[18]/td[2]')\n",
        "    download.click()\n",
        "    print(\"Clicked the second tab!\")\n",
        "except TimeoutException:\n",
        "    print (\"Timed out waiting for page to load\")\n",
        "\n",
        "driver.close()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Clicked the first button!\n",
            "Timed out waiting for page to load\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "E5iGItfImLVG"
      },
      "source": [
        "However, this method ultimately did not work as the second table did not appear after clicking the first click. I was unable to make more progress due to my limited knowledge of HTML and JavaScript. Ultimately, I postulates that there are safeguards put in place to limit any automated downloading/conceal the download as they are an academic institute and may be afraid of request overload. I moved on by downloading the data manually from https://hive.biochemistry.gwu.edu/dna.cgi?cmd=tissue_codon_usage&id=586358&mode=cocoputs (as of 13th of March). "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6MTGgcBhmLVH"
      },
      "source": [
        "## Data exploration\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "Here, we will look at how many entries are there in the file. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "LAdkFzPTmLVH",
        "outputId": "8c9715b6-2eaa-4da4-a0c5-0cee51648963"
      },
      "source": [
        "import pandas as pd\n",
        "#read in CoCoPUTs file\n",
        "df = pd.read_table(\"data/o586358-Refseq_species.tsv\",index_col=False,low_memory=False)\n",
        "print(df.shape)\n",
        "df.head()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "(191025, 76)\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "  Division         Assembly    Taxid                                Species  \\\n",
              "0   refseq  GCF_002287175.1     2161              Methanobacterium bryantii   \n",
              "1   refseq  GCF_000302455.1  1204725   Methanobacterium formicicum DSM 3637   \n",
              "2   refseq  GCF_000762265.1     2162            Methanobacterium formicicum   \n",
              "3   refseq  GCF_001458655.1     2162            Methanobacterium formicicum   \n",
              "4   refseq  GCF_000016525.1   420247  Methanobrevibacter smithii ATCC 35061   \n",
              "\n",
              "  Organelle  Translation Table  # CDS  # Codons    GC%   GC1%  ...   CGA  \\\n",
              "0   genomic                 11   3266    910953  35.52  45.52  ...  2404   \n",
              "1   genomic                 11   2520    719599  40.88  49.69  ...  2968   \n",
              "2   genomic                 11   2376    678104  43.72  51.53  ...  3067   \n",
              "3   genomic                 11   2372    690178  43.50  51.30  ...  3102   \n",
              "4   genomic                 11   1757    557617  32.06  43.34  ...   405   \n",
              "\n",
              "    CGG    AGT   AGC    AGA   AGG    GGT   GGC    GGA   GGG  \n",
              "0  1515  12496  7687  14043  8418  18911  8417  26631  8789  \n",
              "1  3068  10562  4882   8084  6214  17552  5296  21059  7069  \n",
              "2  5204   9335  4793   6499  5439  15945  5847  19130  7839  \n",
              "3  5182   9545  4797   6613  5472  16184  5887  19385  7934  \n",
              "4   168   8234  3248  10191  2880  14664  2965  16981  2107  \n",
              "\n",
              "[5 rows x 76 columns]"
            ],
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Division</th>\n",
              "      <th>Assembly</th>\n",
              "      <th>Taxid</th>\n",
              "      <th>Species</th>\n",
              "      <th>Organelle</th>\n",
              "      <th>Translation Table</th>\n",
              "      <th># CDS</th>\n",
              "      <th># Codons</th>\n",
              "      <th>GC%</th>\n",
              "      <th>GC1%</th>\n",
              "      <th>...</th>\n",
              "      <th>CGA</th>\n",
              "      <th>CGG</th>\n",
              "      <th>AGT</th>\n",
              "      <th>AGC</th>\n",
              "      <th>AGA</th>\n",
              "      <th>AGG</th>\n",
              "      <th>GGT</th>\n",
              "      <th>GGC</th>\n",
              "      <th>GGA</th>\n",
              "      <th>GGG</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_002287175.1</td>\n",
              "      <td>2161</td>\n",
              "      <td>Methanobacterium bryantii</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>3266</td>\n",
              "      <td>910953</td>\n",
              "      <td>35.52</td>\n",
              "      <td>45.52</td>\n",
              "      <td>...</td>\n",
              "      <td>2404</td>\n",
              "      <td>1515</td>\n",
              "      <td>12496</td>\n",
              "      <td>7687</td>\n",
              "      <td>14043</td>\n",
              "      <td>8418</td>\n",
              "      <td>18911</td>\n",
              "      <td>8417</td>\n",
              "      <td>26631</td>\n",
              "      <td>8789</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_000302455.1</td>\n",
              "      <td>1204725</td>\n",
              "      <td>Methanobacterium formicicum DSM 3637</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>2520</td>\n",
              "      <td>719599</td>\n",
              "      <td>40.88</td>\n",
              "      <td>49.69</td>\n",
              "      <td>...</td>\n",
              "      <td>2968</td>\n",
              "      <td>3068</td>\n",
              "      <td>10562</td>\n",
              "      <td>4882</td>\n",
              "      <td>8084</td>\n",
              "      <td>6214</td>\n",
              "      <td>17552</td>\n",
              "      <td>5296</td>\n",
              "      <td>21059</td>\n",
              "      <td>7069</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_000762265.1</td>\n",
              "      <td>2162</td>\n",
              "      <td>Methanobacterium formicicum</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>2376</td>\n",
              "      <td>678104</td>\n",
              "      <td>43.72</td>\n",
              "      <td>51.53</td>\n",
              "      <td>...</td>\n",
              "      <td>3067</td>\n",
              "      <td>5204</td>\n",
              "      <td>9335</td>\n",
              "      <td>4793</td>\n",
              "      <td>6499</td>\n",
              "      <td>5439</td>\n",
              "      <td>15945</td>\n",
              "      <td>5847</td>\n",
              "      <td>19130</td>\n",
              "      <td>7839</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_001458655.1</td>\n",
              "      <td>2162</td>\n",
              "      <td>Methanobacterium formicicum</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>2372</td>\n",
              "      <td>690178</td>\n",
              "      <td>43.50</td>\n",
              "      <td>51.30</td>\n",
              "      <td>...</td>\n",
              "      <td>3102</td>\n",
              "      <td>5182</td>\n",
              "      <td>9545</td>\n",
              "      <td>4797</td>\n",
              "      <td>6613</td>\n",
              "      <td>5472</td>\n",
              "      <td>16184</td>\n",
              "      <td>5887</td>\n",
              "      <td>19385</td>\n",
              "      <td>7934</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_000016525.1</td>\n",
              "      <td>420247</td>\n",
              "      <td>Methanobrevibacter smithii ATCC 35061</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>1757</td>\n",
              "      <td>557617</td>\n",
              "      <td>32.06</td>\n",
              "      <td>43.34</td>\n",
              "      <td>...</td>\n",
              "      <td>405</td>\n",
              "      <td>168</td>\n",
              "      <td>8234</td>\n",
              "      <td>3248</td>\n",
              "      <td>10191</td>\n",
              "      <td>2880</td>\n",
              "      <td>14664</td>\n",
              "      <td>2965</td>\n",
              "      <td>16981</td>\n",
              "      <td>2107</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 76 columns</p>\n",
              "</div>"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 1
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Zd-l5SPxmLVI",
        "outputId": "cca42443-f863-4c3f-9d84-a712d7ecf702"
      },
      "source": [
        "print(list(df))\n",
        "print(df.iloc[:, 0:6].nunique())\n",
        "df.iloc[:,4].unique()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "['Division', 'Assembly', 'Taxid', 'Species', 'Organelle', 'Translation Table', '# CDS', '# Codons', 'GC%', 'GC1%', 'GC2%', 'GC3%', 'TTT', 'TTC', 'TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG', 'ATT', 'ATC', 'ATA', 'ATG', 'GTT', 'GTC', 'GTA', 'GTG', 'TAT', 'TAC', 'TAA', 'TAG', 'CAT', 'CAC', 'CAA', 'CAG', 'AAT', 'AAC', 'AAA', 'AAG', 'GAT', 'GAC', 'GAA', 'GAG', 'TCT', 'TCC', 'TCA', 'TCG', 'CCT', 'CCC', 'CCA', 'CCG', 'ACT', 'ACC', 'ACA', 'ACG', 'GCT', 'GCC', 'GCA', 'GCG', 'TGT', 'TGC', 'TGA', 'TGG', 'CGT', 'CGC', 'CGA', 'CGG', 'AGT', 'AGC', 'AGA', 'AGG', 'GGT', 'GGC', 'GGA', 'GGG']\n",
            "Division                  1\n",
            "Assembly             177439\n",
            "Taxid                 59670\n",
            "Species               59060\n",
            "Organelle                 4\n",
            "Translation Table        11\n",
            "dtype: int64\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['genomic', 'mitochondrion', 'chloroplast', 'plastid'], dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 2
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Xl-9mbXYmLVI"
      },
      "source": [
        "As we can see here, we have 191025 samples with 76 variables. A quick glance at the number of unique values show that the data requires cleaning. Firstly, we have less unique assemblies than number of rows, meaning that some of the assemblies are repeated or NaN. Even though the number of species and taxid differs, here it is acceptable as there might taxonomic IDs only goes up till species levels but there might be subspecies especially for microbes. We also have different type of DNA origins, here we focus on genomic DNA as those genes in mitochondrion, plastid or chloroplast might be shared between distantly related species and skew the differences. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "uXtMNslJmLVJ",
        "outputId": "85016370-7941-4039-bd61-2fa189d1f404"
      },
      "source": [
        "#show duplicated entries\n",
        "df[df.duplicated(['Assembly'])]"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "       Division         Assembly    Taxid                             Species  \\\n",
              "658      refseq  GCF_000513435.1  1173487                 Halorubrum sp. AJ67   \n",
              "660      refseq  GCF_000513435.1  1173487                 Halorubrum sp. AJ67   \n",
              "719      refseq              NaN  1407055  Candidatus Nitrosotenuis uzonensis   \n",
              "720      refseq  GCF_000723185.1  1407055  Candidatus Nitrosotenuis uzonensis   \n",
              "4956     refseq              NaN  1354719          Helicobacter pylori HP87P7   \n",
              "...         ...              ...      ...                                 ...   \n",
              "191020   refseq              NaN  1923763           Xingshan nematode virus 4   \n",
              "191021   refseq              NaN  1923772            Xinzhou nematode virus 4   \n",
              "191022   refseq              NaN  1415147                 Xylella phage Salvo   \n",
              "191023   refseq              NaN  1415148                  Xylella phage Sano   \n",
              "191024   refseq              NaN  2052742          Yersinia phage fHe-Yen9-04   \n",
              "\n",
              "       Organelle  Translation Table  # CDS  # Codons    GC%   GC1%  ...   CGA  \\\n",
              "658      genomic                 11   3078    763055  64.68  67.31  ...  8366   \n",
              "660      genomic                 11   3083    764041  64.69  67.32  ...  8369   \n",
              "719      genomic                 11      1       200  55.83  60.50  ...     1   \n",
              "720      genomic                 11   1876    493394  42.86  49.86  ...  2243   \n",
              "4956     genomic                 11      4      1064  37.53  40.04  ...    11   \n",
              "...          ...                ...    ...       ...    ...    ...  ...   ...   \n",
              "191020   genomic                  0      5      3630  38.21  43.91  ...    49   \n",
              "191021   genomic                  0      5      3652  39.91  45.29  ...    39   \n",
              "191022   genomic                 11     72     17848  63.15  62.21  ...    58   \n",
              "191023   genomic                 11     77     17851  62.46  61.59  ...    71   \n",
              "191024   genomic                 11    531    108127  31.57  43.67  ...   472   \n",
              "\n",
              "          CGG   AGT    AGC   AGA   AGG   GGT    GGC    GGA    GGG  \n",
              "658     17661  3223  10213  1988  2032  7707  29947   8715  16798  \n",
              "660     17683  3223  10222  1988  2032  7711  30002   8725  16820  \n",
              "719         0     1      1     2     2     0      2      5      0  \n",
              "720       824  3560   5730  9304  5257  6294  10132  12660   3397  \n",
              "4956        8    10     22    21    21     9     10     14     16  \n",
              "...       ...   ...    ...   ...   ...   ...    ...    ...    ...  \n",
              "191020     13    60     29    72    21    40     19     85     41  \n",
              "191021     11    71     30    73    12    38     21     99     34  \n",
              "191022    140    30    344    10    22   163    993     71    109  \n",
              "191023    137    41    340    10    24   185    993     74     95  \n",
              "191024    138  2424    329  1074   190  3630    317   1761    401  \n",
              "\n",
              "[13585 rows x 76 columns]"
            ],
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Division</th>\n",
              "      <th>Assembly</th>\n",
              "      <th>Taxid</th>\n",
              "      <th>Species</th>\n",
              "      <th>Organelle</th>\n",
              "      <th>Translation Table</th>\n",
              "      <th># CDS</th>\n",
              "      <th># Codons</th>\n",
              "      <th>GC%</th>\n",
              "      <th>GC1%</th>\n",
              "      <th>...</th>\n",
              "      <th>CGA</th>\n",
              "      <th>CGG</th>\n",
              "      <th>AGT</th>\n",
              "      <th>AGC</th>\n",
              "      <th>AGA</th>\n",
              "      <th>AGG</th>\n",
              "      <th>GGT</th>\n",
              "      <th>GGC</th>\n",
              "      <th>GGA</th>\n",
              "      <th>GGG</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>658</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_000513435.1</td>\n",
              "      <td>1173487</td>\n",
              "      <td>Halorubrum sp. AJ67</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>3078</td>\n",
              "      <td>763055</td>\n",
              "      <td>64.68</td>\n",
              "      <td>67.31</td>\n",
              "      <td>...</td>\n",
              "      <td>8366</td>\n",
              "      <td>17661</td>\n",
              "      <td>3223</td>\n",
              "      <td>10213</td>\n",
              "      <td>1988</td>\n",
              "      <td>2032</td>\n",
              "      <td>7707</td>\n",
              "      <td>29947</td>\n",
              "      <td>8715</td>\n",
              "      <td>16798</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>660</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_000513435.1</td>\n",
              "      <td>1173487</td>\n",
              "      <td>Halorubrum sp. AJ67</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>3083</td>\n",
              "      <td>764041</td>\n",
              "      <td>64.69</td>\n",
              "      <td>67.32</td>\n",
              "      <td>...</td>\n",
              "      <td>8369</td>\n",
              "      <td>17683</td>\n",
              "      <td>3223</td>\n",
              "      <td>10222</td>\n",
              "      <td>1988</td>\n",
              "      <td>2032</td>\n",
              "      <td>7711</td>\n",
              "      <td>30002</td>\n",
              "      <td>8725</td>\n",
              "      <td>16820</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>719</th>\n",
              "      <td>refseq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1407055</td>\n",
              "      <td>Candidatus Nitrosotenuis uzonensis</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>1</td>\n",
              "      <td>200</td>\n",
              "      <td>55.83</td>\n",
              "      <td>60.50</td>\n",
              "      <td>...</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>5</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>720</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_000723185.1</td>\n",
              "      <td>1407055</td>\n",
              "      <td>Candidatus Nitrosotenuis uzonensis</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>1876</td>\n",
              "      <td>493394</td>\n",
              "      <td>42.86</td>\n",
              "      <td>49.86</td>\n",
              "      <td>...</td>\n",
              "      <td>2243</td>\n",
              "      <td>824</td>\n",
              "      <td>3560</td>\n",
              "      <td>5730</td>\n",
              "      <td>9304</td>\n",
              "      <td>5257</td>\n",
              "      <td>6294</td>\n",
              "      <td>10132</td>\n",
              "      <td>12660</td>\n",
              "      <td>3397</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4956</th>\n",
              "      <td>refseq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1354719</td>\n",
              "      <td>Helicobacter pylori HP87P7</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>4</td>\n",
              "      <td>1064</td>\n",
              "      <td>37.53</td>\n",
              "      <td>40.04</td>\n",
              "      <td>...</td>\n",
              "      <td>11</td>\n",
              "      <td>8</td>\n",
              "      <td>10</td>\n",
              "      <td>22</td>\n",
              "      <td>21</td>\n",
              "      <td>21</td>\n",
              "      <td>9</td>\n",
              "      <td>10</td>\n",
              "      <td>14</td>\n",
              "      <td>16</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>191020</th>\n",
              "      <td>refseq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1923763</td>\n",
              "      <td>Xingshan nematode virus 4</td>\n",
              "      <td>genomic</td>\n",
              "      <td>0</td>\n",
              "      <td>5</td>\n",
              "      <td>3630</td>\n",
              "      <td>38.21</td>\n",
              "      <td>43.91</td>\n",
              "      <td>...</td>\n",
              "      <td>49</td>\n",
              "      <td>13</td>\n",
              "      <td>60</td>\n",
              "      <td>29</td>\n",
              "      <td>72</td>\n",
              "      <td>21</td>\n",
              "      <td>40</td>\n",
              "      <td>19</td>\n",
              "      <td>85</td>\n",
              "      <td>41</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>191021</th>\n",
              "      <td>refseq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1923772</td>\n",
              "      <td>Xinzhou nematode virus 4</td>\n",
              "      <td>genomic</td>\n",
              "      <td>0</td>\n",
              "      <td>5</td>\n",
              "      <td>3652</td>\n",
              "      <td>39.91</td>\n",
              "      <td>45.29</td>\n",
              "      <td>...</td>\n",
              "      <td>39</td>\n",
              "      <td>11</td>\n",
              "      <td>71</td>\n",
              "      <td>30</td>\n",
              "      <td>73</td>\n",
              "      <td>12</td>\n",
              "      <td>38</td>\n",
              "      <td>21</td>\n",
              "      <td>99</td>\n",
              "      <td>34</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>191022</th>\n",
              "      <td>refseq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1415147</td>\n",
              "      <td>Xylella phage Salvo</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>72</td>\n",
              "      <td>17848</td>\n",
              "      <td>63.15</td>\n",
              "      <td>62.21</td>\n",
              "      <td>...</td>\n",
              "      <td>58</td>\n",
              "      <td>140</td>\n",
              "      <td>30</td>\n",
              "      <td>344</td>\n",
              "      <td>10</td>\n",
              "      <td>22</td>\n",
              "      <td>163</td>\n",
              "      <td>993</td>\n",
              "      <td>71</td>\n",
              "      <td>109</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>191023</th>\n",
              "      <td>refseq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1415148</td>\n",
              "      <td>Xylella phage Sano</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>77</td>\n",
              "      <td>17851</td>\n",
              "      <td>62.46</td>\n",
              "      <td>61.59</td>\n",
              "      <td>...</td>\n",
              "      <td>71</td>\n",
              "      <td>137</td>\n",
              "      <td>41</td>\n",
              "      <td>340</td>\n",
              "      <td>10</td>\n",
              "      <td>24</td>\n",
              "      <td>185</td>\n",
              "      <td>993</td>\n",
              "      <td>74</td>\n",
              "      <td>95</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>191024</th>\n",
              "      <td>refseq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>2052742</td>\n",
              "      <td>Yersinia phage fHe-Yen9-04</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>531</td>\n",
              "      <td>108127</td>\n",
              "      <td>31.57</td>\n",
              "      <td>43.67</td>\n",
              "      <td>...</td>\n",
              "      <td>472</td>\n",
              "      <td>138</td>\n",
              "      <td>2424</td>\n",
              "      <td>329</td>\n",
              "      <td>1074</td>\n",
              "      <td>190</td>\n",
              "      <td>3630</td>\n",
              "      <td>317</td>\n",
              "      <td>1761</td>\n",
              "      <td>401</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>13585 rows × 76 columns</p>\n",
              "</div>"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 3
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "769CYoeQmLVK"
      },
      "source": [
        "Interestingly, there are multiple entries with the same assembly ID but differing number of coding sequences (# CDS). This may be due to the different gene open reading fram predictions from different softwares. To simplify the data, we shall only keep the assembly with the largest number of CDS. Furhtermore, we shall remove those without assembly IDs, as they are just draft collections of contigs without proper curated annotations. We also remove those with low number of codons."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "xP675yY8mLVK",
        "outputId": "218fa468-dda7-47b8-d5a2-53b9d7bce5eb"
      },
      "source": [
        "df = df[df[\"Organelle\"] == \"genomic\"] #keep genomic\n",
        "df = df[df[\"Assembly\"].notna()] #drop NaN\n",
        "df = df[df[\"# Codons\"] > 20000] #remove low codon counts\n",
        "df = df.sort_values(\"# CDS\", ascending=False).drop_duplicates(\"Assembly\").sort_index()\n",
        "print(df.shape)\n",
        "df[df.duplicated([\"Assembly\"])]"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "(177437, 76)\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Empty DataFrame\n",
              "Columns: [Division, Assembly, Taxid, Species, Organelle, Translation Table, # CDS, # Codons, GC%, GC1%, GC2%, GC3%, TTT, TTC, TTA, TTG, CTT, CTC, CTA, CTG, ATT, ATC, ATA, ATG, GTT, GTC, GTA, GTG, TAT, TAC, TAA, TAG, CAT, CAC, CAA, CAG, AAT, AAC, AAA, AAG, GAT, GAC, GAA, GAG, TCT, TCC, TCA, TCG, CCT, CCC, CCA, CCG, ACT, ACC, ACA, ACG, GCT, GCC, GCA, GCG, TGT, TGC, TGA, TGG, CGT, CGC, CGA, CGG, AGT, AGC, AGA, AGG, GGT, GGC, GGA, GGG]\n",
              "Index: []\n",
              "\n",
              "[0 rows x 76 columns]"
            ],
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Division</th>\n",
              "      <th>Assembly</th>\n",
              "      <th>Taxid</th>\n",
              "      <th>Species</th>\n",
              "      <th>Organelle</th>\n",
              "      <th>Translation Table</th>\n",
              "      <th># CDS</th>\n",
              "      <th># Codons</th>\n",
              "      <th>GC%</th>\n",
              "      <th>GC1%</th>\n",
              "      <th>...</th>\n",
              "      <th>CGA</th>\n",
              "      <th>CGG</th>\n",
              "      <th>AGT</th>\n",
              "      <th>AGC</th>\n",
              "      <th>AGA</th>\n",
              "      <th>AGG</th>\n",
              "      <th>GGT</th>\n",
              "      <th>GGC</th>\n",
              "      <th>GGA</th>\n",
              "      <th>GGG</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>0 rows × 76 columns</p>\n",
              "</div>"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 4
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "eFsgncjHmLVL"
      },
      "source": [
        "Looks like the data is cleaned! We shall check for any NAs just to be safe.."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "aP-ueZoImLVL",
        "outputId": "9aac95e6-95c2-41d0-e204-96b9c1457026"
      },
      "source": [
        "print(df.isnull().any().any())\n",
        "df.head()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "False\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "  Division         Assembly    Taxid                                Species  \\\n",
              "0   refseq  GCF_002287175.1     2161              Methanobacterium bryantii   \n",
              "1   refseq  GCF_000302455.1  1204725   Methanobacterium formicicum DSM 3637   \n",
              "2   refseq  GCF_000762265.1     2162            Methanobacterium formicicum   \n",
              "3   refseq  GCF_001458655.1     2162            Methanobacterium formicicum   \n",
              "4   refseq  GCF_000016525.1   420247  Methanobrevibacter smithii ATCC 35061   \n",
              "\n",
              "  Organelle  Translation Table  # CDS  # Codons    GC%   GC1%  ...   CGA  \\\n",
              "0   genomic                 11   3266    910953  35.52  45.52  ...  2404   \n",
              "1   genomic                 11   2520    719599  40.88  49.69  ...  2968   \n",
              "2   genomic                 11   2376    678104  43.72  51.53  ...  3067   \n",
              "3   genomic                 11   2372    690178  43.50  51.30  ...  3102   \n",
              "4   genomic                 11   1757    557617  32.06  43.34  ...   405   \n",
              "\n",
              "    CGG    AGT   AGC    AGA   AGG    GGT   GGC    GGA   GGG  \n",
              "0  1515  12496  7687  14043  8418  18911  8417  26631  8789  \n",
              "1  3068  10562  4882   8084  6214  17552  5296  21059  7069  \n",
              "2  5204   9335  4793   6499  5439  15945  5847  19130  7839  \n",
              "3  5182   9545  4797   6613  5472  16184  5887  19385  7934  \n",
              "4   168   8234  3248  10191  2880  14664  2965  16981  2107  \n",
              "\n",
              "[5 rows x 76 columns]"
            ],
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Division</th>\n",
              "      <th>Assembly</th>\n",
              "      <th>Taxid</th>\n",
              "      <th>Species</th>\n",
              "      <th>Organelle</th>\n",
              "      <th>Translation Table</th>\n",
              "      <th># CDS</th>\n",
              "      <th># Codons</th>\n",
              "      <th>GC%</th>\n",
              "      <th>GC1%</th>\n",
              "      <th>...</th>\n",
              "      <th>CGA</th>\n",
              "      <th>CGG</th>\n",
              "      <th>AGT</th>\n",
              "      <th>AGC</th>\n",
              "      <th>AGA</th>\n",
              "      <th>AGG</th>\n",
              "      <th>GGT</th>\n",
              "      <th>GGC</th>\n",
              "      <th>GGA</th>\n",
              "      <th>GGG</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_002287175.1</td>\n",
              "      <td>2161</td>\n",
              "      <td>Methanobacterium bryantii</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>3266</td>\n",
              "      <td>910953</td>\n",
              "      <td>35.52</td>\n",
              "      <td>45.52</td>\n",
              "      <td>...</td>\n",
              "      <td>2404</td>\n",
              "      <td>1515</td>\n",
              "      <td>12496</td>\n",
              "      <td>7687</td>\n",
              "      <td>14043</td>\n",
              "      <td>8418</td>\n",
              "      <td>18911</td>\n",
              "      <td>8417</td>\n",
              "      <td>26631</td>\n",
              "      <td>8789</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_000302455.1</td>\n",
              "      <td>1204725</td>\n",
              "      <td>Methanobacterium formicicum DSM 3637</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>2520</td>\n",
              "      <td>719599</td>\n",
              "      <td>40.88</td>\n",
              "      <td>49.69</td>\n",
              "      <td>...</td>\n",
              "      <td>2968</td>\n",
              "      <td>3068</td>\n",
              "      <td>10562</td>\n",
              "      <td>4882</td>\n",
              "      <td>8084</td>\n",
              "      <td>6214</td>\n",
              "      <td>17552</td>\n",
              "      <td>5296</td>\n",
              "      <td>21059</td>\n",
              "      <td>7069</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_000762265.1</td>\n",
              "      <td>2162</td>\n",
              "      <td>Methanobacterium formicicum</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>2376</td>\n",
              "      <td>678104</td>\n",
              "      <td>43.72</td>\n",
              "      <td>51.53</td>\n",
              "      <td>...</td>\n",
              "      <td>3067</td>\n",
              "      <td>5204</td>\n",
              "      <td>9335</td>\n",
              "      <td>4793</td>\n",
              "      <td>6499</td>\n",
              "      <td>5439</td>\n",
              "      <td>15945</td>\n",
              "      <td>5847</td>\n",
              "      <td>19130</td>\n",
              "      <td>7839</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_001458655.1</td>\n",
              "      <td>2162</td>\n",
              "      <td>Methanobacterium formicicum</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>2372</td>\n",
              "      <td>690178</td>\n",
              "      <td>43.50</td>\n",
              "      <td>51.30</td>\n",
              "      <td>...</td>\n",
              "      <td>3102</td>\n",
              "      <td>5182</td>\n",
              "      <td>9545</td>\n",
              "      <td>4797</td>\n",
              "      <td>6613</td>\n",
              "      <td>5472</td>\n",
              "      <td>16184</td>\n",
              "      <td>5887</td>\n",
              "      <td>19385</td>\n",
              "      <td>7934</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_000016525.1</td>\n",
              "      <td>420247</td>\n",
              "      <td>Methanobrevibacter smithii ATCC 35061</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>1757</td>\n",
              "      <td>557617</td>\n",
              "      <td>32.06</td>\n",
              "      <td>43.34</td>\n",
              "      <td>...</td>\n",
              "      <td>405</td>\n",
              "      <td>168</td>\n",
              "      <td>8234</td>\n",
              "      <td>3248</td>\n",
              "      <td>10191</td>\n",
              "      <td>2880</td>\n",
              "      <td>14664</td>\n",
              "      <td>2965</td>\n",
              "      <td>16981</td>\n",
              "      <td>2107</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 76 columns</p>\n",
              "</div>"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 5
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "JVZ4nnpsmLVM"
      },
      "source": [
        "## Preparing data for training and testing\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "We will split the data up, using the codons as variables (total of 64) and using the taxonomy IDs as class labels. We will need to parse the taxonomy tree for the lineage of each taxonomy id. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "J875AptCmLVN",
        "outputId": "7f8b5c8a-9410-4a16-ddf6-20e6c2544bff"
      },
      "source": [
        "from Bio import Entrez\n",
        "#entrez records has maximum length of 10000\n",
        "total_species = len(df.iloc[:, 2].unique())\n",
        "i=0\n",
        "taxid = []\n",
        "while i < total_species:\n",
        "    if i == 50000:\n",
        "        i_end = total_species\n",
        "    else:\n",
        "        i_end = i+10000\n",
        "    handle = Entrez.efetch(db=\"Taxonomy\", id=df.iloc[:, 2].unique()[i:i_end], retmode=\"xml\")\n",
        "    records = Entrez.read(handle)\n",
        "    taxid.append(records)\n",
        "    i+=10000"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "C:\\Users\\YZ\\miniconda3\\envs\\project\\lib\\site-packages\\Bio\\Entrez\\__init__.py:658: UserWarning: \n",
            "Email address is not specified.\n",
            "\n",
            "To make use of NCBI's E-utilities, NCBI requires you to specify your\n",
            "email address with each request.  As an example, if your email address\n",
            "is A.N.Other@example.com, you can specify it as follows:\n",
            "   from Bio import Entrez\n",
            "   Entrez.email = 'A.N.Other@example.com'\n",
            "In case of excessive usage of the E-utilities, NCBI will attempt to contact\n",
            "a user at the email address provided before blocking access to the\n",
            "E-utilities.\n",
            "  warnings.warn(\n"
          ],
          "name": "stderr"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "tags": [],
        "id": "zZgOAPXdmLVN"
      },
      "source": [
        "#create empty taxa table to keep track of each taxa lineage\n",
        "#0 is a safe value as 0 correspond to the root node of the taxanomy tree at NCBI, which is a non valid annotation\n",
        "tax_df = pd.DataFrame(0, index=range(total_species), columns=range(7))\n",
        "major_taxonomy_levels = [\"superkingdom\", \"phylum\", \"class\", \"order\", \"family\", \"genus\", \"Taxid\"]\n",
        "tax_df.columns = major_taxonomy_levels"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "tags": [],
        "id": "i9774nCGmLVO",
        "outputId": "09c6f506-b849-4313-97ea-08bc6468334f"
      },
      "source": [
        "#populate the taxa table for each record\n",
        "i=0\n",
        "for records in taxid:\n",
        "    for record in records:\n",
        "        tax_df.iloc[i,6] = int(record[\"TaxId\"])\n",
        "        for tax in record[\"LineageEx\"]:\n",
        "            if tax[\"Rank\"] == \"superkingdom\":\n",
        "                tax_df.iloc[i,0] = int(tax[\"TaxId\"])\n",
        "            elif tax[\"Rank\"] == \"phylum\":\n",
        "                tax_df.iloc[i,1] = int(tax[\"TaxId\"])\n",
        "            elif tax[\"Rank\"] == \"class\":\n",
        "                tax_df.iloc[i,2] = int(tax[\"TaxId\"])\n",
        "            elif tax[\"Rank\"] == \"order\":\n",
        "                tax_df.iloc[i,3] = int(tax[\"TaxId\"])\n",
        "            elif tax[\"Rank\"] == \"family\":\n",
        "                tax_df.iloc[i,4] = int(tax[\"TaxId\"])\n",
        "            elif tax[\"Rank\"] == \"genus\":\n",
        "                tax_df.iloc[i,5] = int(tax[\"TaxId\"])\n",
        "        i+=1\n",
        "tax_df       "
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "       superkingdom  phylum    class  order   family   genus    Taxid\n",
              "0              2157   28890   183925   2158     2159    2160     2161\n",
              "1              2157   28890   183925   2158     2159    2160  1204725\n",
              "2              2157   28890   183925   2158     2159    2160     2162\n",
              "3              2157   28890   183925   2158     2159    2172   420247\n",
              "4              2157   28890   183925   2158     2159    2172   521002\n",
              "...             ...     ...      ...    ...      ...     ...      ...\n",
              "50392          2759    7711     8292   8445  1277737  264009  1415580\n",
              "50393          2759    7711   186623   7952     7953   75365  1608454\n",
              "50394          2759    7711   186623  41712    31092   91732  1676925\n",
              "50395          2759    7711  1338369   8459     8487   38771  1825980\n",
              "50396          2759    7711     8782   9223     9224   35547  2489341\n",
              "\n",
              "[50397 rows x 7 columns]"
            ],
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>superkingdom</th>\n",
              "      <th>phylum</th>\n",
              "      <th>class</th>\n",
              "      <th>order</th>\n",
              "      <th>family</th>\n",
              "      <th>genus</th>\n",
              "      <th>Taxid</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>2157</td>\n",
              "      <td>28890</td>\n",
              "      <td>183925</td>\n",
              "      <td>2158</td>\n",
              "      <td>2159</td>\n",
              "      <td>2160</td>\n",
              "      <td>2161</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>2157</td>\n",
              "      <td>28890</td>\n",
              "      <td>183925</td>\n",
              "      <td>2158</td>\n",
              "      <td>2159</td>\n",
              "      <td>2160</td>\n",
              "      <td>1204725</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2157</td>\n",
              "      <td>28890</td>\n",
              "      <td>183925</td>\n",
              "      <td>2158</td>\n",
              "      <td>2159</td>\n",
              "      <td>2160</td>\n",
              "      <td>2162</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>2157</td>\n",
              "      <td>28890</td>\n",
              "      <td>183925</td>\n",
              "      <td>2158</td>\n",
              "      <td>2159</td>\n",
              "      <td>2172</td>\n",
              "      <td>420247</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>2157</td>\n",
              "      <td>28890</td>\n",
              "      <td>183925</td>\n",
              "      <td>2158</td>\n",
              "      <td>2159</td>\n",
              "      <td>2172</td>\n",
              "      <td>521002</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>50392</th>\n",
              "      <td>2759</td>\n",
              "      <td>7711</td>\n",
              "      <td>8292</td>\n",
              "      <td>8445</td>\n",
              "      <td>1277737</td>\n",
              "      <td>264009</td>\n",
              "      <td>1415580</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>50393</th>\n",
              "      <td>2759</td>\n",
              "      <td>7711</td>\n",
              "      <td>186623</td>\n",
              "      <td>7952</td>\n",
              "      <td>7953</td>\n",
              "      <td>75365</td>\n",
              "      <td>1608454</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>50394</th>\n",
              "      <td>2759</td>\n",
              "      <td>7711</td>\n",
              "      <td>186623</td>\n",
              "      <td>41712</td>\n",
              "      <td>31092</td>\n",
              "      <td>91732</td>\n",
              "      <td>1676925</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>50395</th>\n",
              "      <td>2759</td>\n",
              "      <td>7711</td>\n",
              "      <td>1338369</td>\n",
              "      <td>8459</td>\n",
              "      <td>8487</td>\n",
              "      <td>38771</td>\n",
              "      <td>1825980</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>50396</th>\n",
              "      <td>2759</td>\n",
              "      <td>7711</td>\n",
              "      <td>8782</td>\n",
              "      <td>9223</td>\n",
              "      <td>9224</td>\n",
              "      <td>35547</td>\n",
              "      <td>2489341</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>50397 rows × 7 columns</p>\n",
              "</div>"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "cBBKDgcDmLVO",
        "outputId": "0d3022b3-9259-41fa-ebdc-a38495862f12"
      },
      "source": [
        "print((tax_df == 0).sum(axis=0)) #find number of zeroes in each taxa\n",
        "print(len(tax_df.iloc[:, 1].unique())) #find number of unique taxa at phylum level\n",
        "from collections import Counter\n",
        "Counter(tax_df[\"phylum\"]).values() #check distribution of number records for each phyla"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "superkingdom       0\n",
            "phylum            71\n",
            "class            785\n",
            "order            424\n",
            "family           837\n",
            "genus           1077\n",
            "Taxid              0\n",
            "dtype: int64\n",
            "88\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "dict_values([603, 91, 23, 33, 3, 1, 1, 71, 1, 21866, 8127, 4, 2663, 148, 615, 23, 13256, 116, 74, 183, 134, 43, 11, 32, 26, 547, 354, 83, 9, 132, 29, 50, 3, 1, 1, 5, 4, 3, 1, 15, 2, 6, 2, 2, 22, 1, 5, 1, 3, 1, 4, 1, 1, 2, 1, 3, 1, 1, 6, 4, 210, 4, 52, 8, 3, 9, 4, 4, 1, 9, 143, 1, 2, 309, 1, 1, 1, 1, 108, 8, 1, 2, 1, 7, 13, 4, 35, 1])"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 9
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "D7d9x7I6mLVP"
      },
      "source": [
        "There appears to be some data with missing taxonomic annotations. Furthermore, some phyla has many examples while others only have a handful. We will only keep those with > 100 examples."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "AcR4wG21mLVP",
        "outputId": "0a625c15-7942-4e6a-e565-68c696ebbf6f"
      },
      "source": [
        "#keep taxa with only > 100 samples\n",
        "phylum_to_keep = [key for key,value in Counter(tax_df[\"phylum\"]).items() if value > 100]\n",
        "tax_df = tax_df.loc[tax_df[\"phylum\"].isin(phylum_to_keep)]\n",
        "taxid_to_keep = tax_df[\"Taxid\"]\n",
        "df = df.loc[df[\"Taxid\"].isin(taxid_to_keep)]\n",
        "df"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "       Division         Assembly    Taxid  \\\n",
              "0        refseq  GCF_002287175.1     2161   \n",
              "1        refseq  GCF_000302455.1  1204725   \n",
              "2        refseq  GCF_000762265.1     2162   \n",
              "3        refseq  GCF_001458655.1     2162   \n",
              "4        refseq  GCF_000016525.1   420247   \n",
              "...         ...              ...      ...   \n",
              "181848   refseq  GCF_901765095.1  1415580   \n",
              "181850   refseq  GCF_001515605.1  1608454   \n",
              "181852   refseq  GCF_002872115.1  1676925   \n",
              "181853   refseq  GCF_007399415.2  1825980   \n",
              "181855   refseq  GCF_004027225.1  2489341   \n",
              "\n",
              "                                      Species Organelle  Translation Table  \\\n",
              "0                   Methanobacterium bryantii   genomic                 11   \n",
              "1        Methanobacterium formicicum DSM 3637   genomic                 11   \n",
              "2                 Methanobacterium formicicum   genomic                 11   \n",
              "3                 Methanobacterium formicicum   genomic                 11   \n",
              "4       Methanobrevibacter smithii ATCC 35061   genomic                 11   \n",
              "...                                       ...       ...                ...   \n",
              "181848                 Microcaecilia unicolor   genomic                  0   \n",
              "181850           Sinocyclocheilus anshuiensis   genomic                  0   \n",
              "181852               Paramormyrops kingsleyae   genomic                  0   \n",
              "181853                      Gopherus evgoodei   genomic                  0   \n",
              "181855                    Strigops habroptila   genomic                  0   \n",
              "\n",
              "        # CDS  # Codons    GC%   GC1%  ...     CGA     CGG     AGT     AGC  \\\n",
              "0        3266    910953  35.52  45.52  ...    2404    1515   12496    7687   \n",
              "1        2520    719599  40.88  49.69  ...    2968    3068   10562    4882   \n",
              "2        2376    678104  43.72  51.53  ...    3067    5204    9335    4793   \n",
              "3        2372    690178  43.50  51.30  ...    3102    5182    9545    4797   \n",
              "4        1757    557617  32.06  43.34  ...     405     168    8234    3248   \n",
              "...       ...       ...    ...    ...  ...     ...     ...     ...     ...   \n",
              "181848  35405  20605114  46.78  52.18  ...  137037  136305  330959  359823   \n",
              "181850  67753  41525357  50.84  54.62  ...  266568  284411  622434  819038   \n",
              "181852  55174  37697302  56.10  57.17  ...  189905  463292  403279  918378   \n",
              "181853  49709  34253340  49.30  54.06  ...  199608  300008  500034  677088   \n",
              "181855  42710  31775684  49.39  54.52  ...  178349  285596  463765  640730   \n",
              "\n",
              "           AGA     AGG     GGT     GGC     GGA     GGG  \n",
              "0        14043    8418   18911    8417   26631    8789  \n",
              "1         8084    6214   17552    5296   21059    7069  \n",
              "2         6499    5439   15945    5847   19130    7839  \n",
              "3         6613    5472   16184    5887   19385    7934  \n",
              "4        10191    2880   14664    2965   16981    2107  \n",
              "...        ...     ...     ...     ...     ...     ...  \n",
              "181848  321754  242704  264062  274277  410074  241232  \n",
              "181850  615112  470561  550460  697798  812809  451550  \n",
              "181852  335427  566434  388203  913712  520100  646555  \n",
              "181853  516372  422769  365846  548610  609307  517676  \n",
              "181855  453429  390137  377436  520866  590895  448712  \n",
              "\n",
              "[174160 rows x 76 columns]"
            ],
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Division</th>\n",
              "      <th>Assembly</th>\n",
              "      <th>Taxid</th>\n",
              "      <th>Species</th>\n",
              "      <th>Organelle</th>\n",
              "      <th>Translation Table</th>\n",
              "      <th># CDS</th>\n",
              "      <th># Codons</th>\n",
              "      <th>GC%</th>\n",
              "      <th>GC1%</th>\n",
              "      <th>...</th>\n",
              "      <th>CGA</th>\n",
              "      <th>CGG</th>\n",
              "      <th>AGT</th>\n",
              "      <th>AGC</th>\n",
              "      <th>AGA</th>\n",
              "      <th>AGG</th>\n",
              "      <th>GGT</th>\n",
              "      <th>GGC</th>\n",
              "      <th>GGA</th>\n",
              "      <th>GGG</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_002287175.1</td>\n",
              "      <td>2161</td>\n",
              "      <td>Methanobacterium bryantii</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>3266</td>\n",
              "      <td>910953</td>\n",
              "      <td>35.52</td>\n",
              "      <td>45.52</td>\n",
              "      <td>...</td>\n",
              "      <td>2404</td>\n",
              "      <td>1515</td>\n",
              "      <td>12496</td>\n",
              "      <td>7687</td>\n",
              "      <td>14043</td>\n",
              "      <td>8418</td>\n",
              "      <td>18911</td>\n",
              "      <td>8417</td>\n",
              "      <td>26631</td>\n",
              "      <td>8789</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_000302455.1</td>\n",
              "      <td>1204725</td>\n",
              "      <td>Methanobacterium formicicum DSM 3637</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>2520</td>\n",
              "      <td>719599</td>\n",
              "      <td>40.88</td>\n",
              "      <td>49.69</td>\n",
              "      <td>...</td>\n",
              "      <td>2968</td>\n",
              "      <td>3068</td>\n",
              "      <td>10562</td>\n",
              "      <td>4882</td>\n",
              "      <td>8084</td>\n",
              "      <td>6214</td>\n",
              "      <td>17552</td>\n",
              "      <td>5296</td>\n",
              "      <td>21059</td>\n",
              "      <td>7069</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_000762265.1</td>\n",
              "      <td>2162</td>\n",
              "      <td>Methanobacterium formicicum</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>2376</td>\n",
              "      <td>678104</td>\n",
              "      <td>43.72</td>\n",
              "      <td>51.53</td>\n",
              "      <td>...</td>\n",
              "      <td>3067</td>\n",
              "      <td>5204</td>\n",
              "      <td>9335</td>\n",
              "      <td>4793</td>\n",
              "      <td>6499</td>\n",
              "      <td>5439</td>\n",
              "      <td>15945</td>\n",
              "      <td>5847</td>\n",
              "      <td>19130</td>\n",
              "      <td>7839</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_001458655.1</td>\n",
              "      <td>2162</td>\n",
              "      <td>Methanobacterium formicicum</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>2372</td>\n",
              "      <td>690178</td>\n",
              "      <td>43.50</td>\n",
              "      <td>51.30</td>\n",
              "      <td>...</td>\n",
              "      <td>3102</td>\n",
              "      <td>5182</td>\n",
              "      <td>9545</td>\n",
              "      <td>4797</td>\n",
              "      <td>6613</td>\n",
              "      <td>5472</td>\n",
              "      <td>16184</td>\n",
              "      <td>5887</td>\n",
              "      <td>19385</td>\n",
              "      <td>7934</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_000016525.1</td>\n",
              "      <td>420247</td>\n",
              "      <td>Methanobrevibacter smithii ATCC 35061</td>\n",
              "      <td>genomic</td>\n",
              "      <td>11</td>\n",
              "      <td>1757</td>\n",
              "      <td>557617</td>\n",
              "      <td>32.06</td>\n",
              "      <td>43.34</td>\n",
              "      <td>...</td>\n",
              "      <td>405</td>\n",
              "      <td>168</td>\n",
              "      <td>8234</td>\n",
              "      <td>3248</td>\n",
              "      <td>10191</td>\n",
              "      <td>2880</td>\n",
              "      <td>14664</td>\n",
              "      <td>2965</td>\n",
              "      <td>16981</td>\n",
              "      <td>2107</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>181848</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_901765095.1</td>\n",
              "      <td>1415580</td>\n",
              "      <td>Microcaecilia unicolor</td>\n",
              "      <td>genomic</td>\n",
              "      <td>0</td>\n",
              "      <td>35405</td>\n",
              "      <td>20605114</td>\n",
              "      <td>46.78</td>\n",
              "      <td>52.18</td>\n",
              "      <td>...</td>\n",
              "      <td>137037</td>\n",
              "      <td>136305</td>\n",
              "      <td>330959</td>\n",
              "      <td>359823</td>\n",
              "      <td>321754</td>\n",
              "      <td>242704</td>\n",
              "      <td>264062</td>\n",
              "      <td>274277</td>\n",
              "      <td>410074</td>\n",
              "      <td>241232</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>181850</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_001515605.1</td>\n",
              "      <td>1608454</td>\n",
              "      <td>Sinocyclocheilus anshuiensis</td>\n",
              "      <td>genomic</td>\n",
              "      <td>0</td>\n",
              "      <td>67753</td>\n",
              "      <td>41525357</td>\n",
              "      <td>50.84</td>\n",
              "      <td>54.62</td>\n",
              "      <td>...</td>\n",
              "      <td>266568</td>\n",
              "      <td>284411</td>\n",
              "      <td>622434</td>\n",
              "      <td>819038</td>\n",
              "      <td>615112</td>\n",
              "      <td>470561</td>\n",
              "      <td>550460</td>\n",
              "      <td>697798</td>\n",
              "      <td>812809</td>\n",
              "      <td>451550</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>181852</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_002872115.1</td>\n",
              "      <td>1676925</td>\n",
              "      <td>Paramormyrops kingsleyae</td>\n",
              "      <td>genomic</td>\n",
              "      <td>0</td>\n",
              "      <td>55174</td>\n",
              "      <td>37697302</td>\n",
              "      <td>56.10</td>\n",
              "      <td>57.17</td>\n",
              "      <td>...</td>\n",
              "      <td>189905</td>\n",
              "      <td>463292</td>\n",
              "      <td>403279</td>\n",
              "      <td>918378</td>\n",
              "      <td>335427</td>\n",
              "      <td>566434</td>\n",
              "      <td>388203</td>\n",
              "      <td>913712</td>\n",
              "      <td>520100</td>\n",
              "      <td>646555</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>181853</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_007399415.2</td>\n",
              "      <td>1825980</td>\n",
              "      <td>Gopherus evgoodei</td>\n",
              "      <td>genomic</td>\n",
              "      <td>0</td>\n",
              "      <td>49709</td>\n",
              "      <td>34253340</td>\n",
              "      <td>49.30</td>\n",
              "      <td>54.06</td>\n",
              "      <td>...</td>\n",
              "      <td>199608</td>\n",
              "      <td>300008</td>\n",
              "      <td>500034</td>\n",
              "      <td>677088</td>\n",
              "      <td>516372</td>\n",
              "      <td>422769</td>\n",
              "      <td>365846</td>\n",
              "      <td>548610</td>\n",
              "      <td>609307</td>\n",
              "      <td>517676</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>181855</th>\n",
              "      <td>refseq</td>\n",
              "      <td>GCF_004027225.1</td>\n",
              "      <td>2489341</td>\n",
              "      <td>Strigops habroptila</td>\n",
              "      <td>genomic</td>\n",
              "      <td>0</td>\n",
              "      <td>42710</td>\n",
              "      <td>31775684</td>\n",
              "      <td>49.39</td>\n",
              "      <td>54.52</td>\n",
              "      <td>...</td>\n",
              "      <td>178349</td>\n",
              "      <td>285596</td>\n",
              "      <td>463765</td>\n",
              "      <td>640730</td>\n",
              "      <td>453429</td>\n",
              "      <td>390137</td>\n",
              "      <td>377436</td>\n",
              "      <td>520866</td>\n",
              "      <td>590895</td>\n",
              "      <td>448712</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>174160 rows × 76 columns</p>\n",
              "</div>"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 10
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ijFe1-ptmLVQ"
      },
      "source": [
        "#change to percentage\n",
        "df.iloc[:,12:76] = df.iloc[:,12:76].div(df[\"# Codons\"].values, axis=0)\n",
        "#add phylum data\n",
        "df = df.merge(tax_df[[\"phylum\", \"Taxid\"]], how=\"left\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "EkUNzZygNRi2"
      },
      "source": [
        "We shall save the data for future convenience."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "aGF1kFg_mLVR"
      },
      "source": [
        "#split into X and y data\n",
        "X = df.iloc[:, 12:76]\n",
        "y = df.iloc[:, 76]\n",
        "#save data\n",
        "X.to_csv(\"data/X_cleaned.csv\")\n",
        "y.to_csv(\"data/y_cleaned.csv\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "sd2LQg2VQZE5"
      },
      "source": [
        "Side note on not normalizing and scaling the data: \n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "Due to the fact that the genome sizes and hence the number of coding sequences vary wildly from single cell organisms such as bacteria to multicellular organisms such as mammmals, I have decided to just use the percentage of each codon the input variables for the machine learning algorithms. However, this leads to a whole class of problems on its own, a field termed as compositional data analysis (CoDA). In brief, the constant sum restraint means that each variable is negatively correlated to every other variable. Furthermore, spurious correlations occurs due to the shared denominator between all variables. In statistical analysis, these spurious correlations lead to high number of false positives. In machine learning, it means that some algorithms are not suitable for the task at hand, such as Naive Bayes, which was initially used in this project. One possible solution is to transfrom the data with the class of transformation called log ratio transformation. However, how to best prepare the data and apply these transformations is an active area of research and hence I decided not to proceed with that here. "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "hxBWYa8Covx_"
      },
      "source": [
        "## Training machine learning algorithms\n",
        "\n",
        "---\n",
        "\n",
        "In this section, we will explore 3 different machine learning algorithms for classification. They are the k-Nearest Neighbours, Random Forest, and XGBoost. All methods will be evaluated with the `classification_report` from `sklearn.metrics`, which shows the weighted F1 scores, precision and recall among others. The F1 score, which is the harmonic mean of precision and recall, is especially important in this case with imbalanced data classes. \n",
        "\n",
        "First, we will load in the data cleaned from before."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "wXW-U5RoQVWC"
      },
      "source": [
        "#local machine ver\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "X = pd.read_csv(\"data/X_cleaned.csv\", index_col=0)\n",
        "y = pd.read_csv(\"data/y_cleaned.csv\", index_col=0)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "kSHE8dl2nK6R",
        "outputId": "c7eae17f-c727-4845-9f7b-f8ebf3985bbf"
      },
      "source": [
        "#colab ver\n",
        "from google.colab import drive\n",
        "import pandas as pd\n",
        "\n",
        "drive.mount(\"/content/gdrive\")\n",
        "X = pd.read_csv(\"/content/gdrive/MyDrive/ce9010/X_cleaned.csv\", index_col=0) #replace with path to your file\n",
        "y = pd.read_csv(\"/content/gdrive/MyDrive/ce9010/y_cleaned.csv\", index_col=0) #replace with path to your file"
      ],
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount(\"/content/gdrive\", force_remount=True).\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VDfIcQq-mLVS"
      },
      "source": [
        "### Training k-Nearest Neighbour classifier\n",
        "---\n",
        "k-Nearest Neighbour is a simple algorithm that works on the principle that similar objects have similar characteristics, hence, they lie close to each other in the variable space. The class for an unknown object is assigned based on the most common class of k nearest objects. Here, we will use the default  Euclidean distance metric and vary the k variable for optimal predictive power. We will use the `StratifiedKFold` from `sklearn.model_selection` to find the best k value. \n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8BFij5JELCCn",
        "outputId": "4698c5d7-a8a3-41a9-89a3-99ae1be0c468"
      },
      "source": [
        "from sklearn.neighbors import KNeighborsClassifier\n",
        "from sklearn.model_selection import StratifiedKFold, train_test_split\n",
        "from sklearn import preprocessing\n",
        "from sklearn.metrics import classification_report\n",
        "\n",
        "# rename the classes\n",
        "le = preprocessing.LabelEncoder()\n",
        "y_encoded = le.fit_transform(y[:,0])\n",
        "\n",
        "# knn k choosing\n",
        "X_train, X_test, y_train, y_test = train_test_split(X, y_encoded,stratify=y_encoded,test_size=0.3,random_state=42) #70% training and 30% test\n",
        "skf = StratifiedKFold(n_splits=10)\n",
        "k = 1\n",
        "for train_index, test_index in skf.split(X_train, y_train):\n",
        "    kfold_X_train = X_train[train_index,:]\n",
        "    kfold_y_train = y_train[train_index]\n",
        "    kfold_X_test = X_train[test_index,:]\n",
        "    kfold_y_test = y_train[test_index]\n",
        "    knn = KNeighborsClassifier(n_neighbors=k)\n",
        "    knn.fit(kfold_X_train,kfold_y_train)\n",
        "    preds = knn.predict(kfold_X_test)\n",
        "    print(f\"KNN report for {k} Nearest Neighbours:\")\n",
        "    print(classification_report(kfold_y_test, preds,digits=4))\n",
        "    k+=1"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "KNN report for 1 Nearest Neighbours:\n",
            "              precision    recall  f1-score   support\n",
            "\n",
            "           0     1.0000    0.9905    0.9952       317\n",
            "           1     0.9756    1.0000    0.9877        40\n",
            "           2     0.9987    0.9994    0.9991      7025\n",
            "           3     0.9991    0.9993    0.9992      4573\n",
            "           4     1.0000    1.0000    1.0000        12\n",
            "           5     1.0000    0.9333    0.9655        15\n",
            "           6     1.0000    1.0000    1.0000        10\n",
            "           7     1.0000    1.0000    1.0000        22\n",
            "           8     1.0000    0.9808    0.9903        52\n",
            "           9     1.0000    1.0000    1.0000        18\n",
            "          10     0.8750    1.0000    0.9333         7\n",
            "          11     1.0000    0.8947    0.9444        19\n",
            "          12     0.9984    1.0000    0.9992      1266\n",
            "          13     0.9167    0.9167    0.9167        12\n",
            "          14     1.0000    0.9595    0.9793        74\n",
            "          15     1.0000    1.0000    1.0000        28\n",
            "          16     1.0000    1.0000    1.0000        47\n",
            "\n",
            "    accuracy                         0.9987     13537\n",
            "   macro avg     0.9861    0.9808    0.9829     13537\n",
            "weighted avg     0.9987    0.9987    0.9987     13537\n",
            "\n",
            "KNN report for 2 Nearest Neighbours:\n",
            "              precision    recall  f1-score   support\n",
            "\n",
            "           0     0.9906    0.9968    0.9937       317\n",
            "           1     0.9750    1.0000    0.9873        39\n",
            "           2     0.9989    0.9994    0.9991      7024\n",
            "           3     0.9996    0.9993    0.9995      4573\n",
            "           4     1.0000    1.0000    1.0000        12\n",
            "           5     0.8235    1.0000    0.9032        14\n",
            "           6     1.0000    0.8000    0.8889        10\n",
            "           7     1.0000    1.0000    1.0000        22\n",
            "           8     1.0000    0.9808    0.9903        52\n",
            "           9     1.0000    1.0000    1.0000        18\n",
            "          10     1.0000    1.0000    1.0000         8\n",
            "          11     1.0000    0.8889    0.9412        18\n",
            "          12     0.9961    0.9984    0.9972      1267\n",
            "          13     1.0000    0.6154    0.7619        13\n",
            "          14     1.0000    0.9730    0.9863        74\n",
            "          15     1.0000    1.0000    1.0000        28\n",
            "          16     1.0000    1.0000    1.0000        48\n",
            "\n",
            "    accuracy                         0.9984     13537\n",
            "   macro avg     0.9873    0.9560    0.9676     13537\n",
            "weighted avg     0.9984    0.9984    0.9983     13537\n",
            "\n",
            "KNN report for 3 Nearest Neighbours:\n",
            "              precision    recall  f1-score   support\n",
            "\n",
            "           0     0.9937    0.9968    0.9953       317\n",
            "           1     1.0000    0.9744    0.9870        39\n",
            "           2     0.9974    0.9996    0.9985      7024\n",
            "           3     0.9983    0.9985    0.9984      4573\n",
            "           4     1.0000    1.0000    1.0000        12\n",
            "           5     1.0000    1.0000    1.0000        14\n",
            "           6     1.0000    1.0000    1.0000        10\n",
            "           7     1.0000    1.0000    1.0000        22\n",
            "           8     1.0000    0.9434    0.9709        53\n",
            "           9     0.9000    1.0000    0.9474        18\n",
            "          10     1.0000    1.0000    1.0000         8\n",
            "          11     1.0000    0.7778    0.8750        18\n",
            "          12     0.9961    0.9968    0.9964      1267\n",
            "          13     1.0000    0.3846    0.5556        13\n",
            "          14     1.0000    0.9589    0.9790        73\n",
            "          15     1.0000    1.0000    1.0000        28\n",
            "          16     1.0000    0.9792    0.9895        48\n",
            "\n",
            "    accuracy                         0.9974     13537\n",
            "   macro avg     0.9933    0.9418    0.9584     13537\n",
            "weighted avg     0.9974    0.9974    0.9973     13537\n",
            "\n",
            "KNN report for 4 Nearest Neighbours:\n",
            "              precision    recall  f1-score   support\n",
            "\n",
            "           0     0.9843    0.9905    0.9874       317\n",
            "           1     0.9744    0.9500    0.9620        40\n",
            "           2     0.9984    0.9991    0.9988      7024\n",
            "           3     0.9987    0.9991    0.9989      4573\n",
            "           4     1.0000    1.0000    1.0000        12\n",
            "           5     0.8750    1.0000    0.9333        14\n",
            "           6     1.0000    0.8000    0.8889        10\n",
            "           7     0.9565    1.0000    0.9778        22\n",
            "           8     0.9808    0.9808    0.9808        52\n",
            "           9     1.0000    1.0000    1.0000        18\n",
            "          10     1.0000    1.0000    1.0000         8\n",
            "          11     1.0000    0.8947    0.9444        19\n",
            "          12     0.9968    0.9968    0.9968      1267\n",
            "          13     1.0000    0.6154    0.7619        13\n",
            "          14     1.0000    0.9730    0.9863        74\n",
            "          15     1.0000    1.0000    1.0000        27\n",
            "          16     1.0000    1.0000    1.0000        47\n",
            "\n",
            "    accuracy                         0.9977     13537\n",
            "   macro avg     0.9862    0.9529    0.9657     13537\n",
            "weighted avg     0.9977    0.9977    0.9977     13537\n",
            "\n",
            "KNN report for 5 Nearest Neighbours:\n",
            "              precision    recall  f1-score   support\n",
            "\n",
            "           0     0.9937    0.9968    0.9953       317\n",
            "           1     1.0000    0.9500    0.9744        40\n",
            "           2     0.9979    0.9993    0.9986      7024\n",
            "           3     0.9993    0.9989    0.9991      4573\n",
            "           4     1.0000    1.0000    1.0000        12\n",
            "           5     0.9375    1.0000    0.9677        15\n",
            "           6     1.0000    0.9000    0.9474        10\n",
            "           7     1.0000    1.0000    1.0000        22\n",
            "           8     0.9811    1.0000    0.9905        52\n",
            "           9     0.9375    0.8824    0.9091        17\n",
            "          10     0.8000    1.0000    0.8889         8\n",
            "          11     1.0000    0.7895    0.8824        19\n",
            "          12     0.9976    0.9968    0.9972      1267\n",
            "          13     0.8889    0.6667    0.7619        12\n",
            "          14     1.0000    1.0000    1.0000        74\n",
            "          15     1.0000    0.9630    0.9811        27\n",
            "          16     1.0000    1.0000    1.0000        47\n",
            "\n",
            "    accuracy                         0.9979     13536\n",
            "   macro avg     0.9726    0.9496    0.9584     13536\n",
            "weighted avg     0.9979    0.9979    0.9978     13536\n",
            "\n",
            "KNN report for 6 Nearest Neighbours:\n",
            "              precision    recall  f1-score   support\n",
            "\n",
            "           0     1.0000    0.9905    0.9952       317\n",
            "           1     0.9756    1.0000    0.9877        40\n",
            "           2     0.9967    0.9994    0.9981      7024\n",
            "           3     0.9974    0.9991    0.9983      4573\n",
            "           4     1.0000    0.6364    0.7778        11\n",
            "           5     1.0000    0.8000    0.8889        15\n",
            "           6     0.9091    1.0000    0.9524        10\n",
            "           7     1.0000    1.0000    1.0000        22\n",
            "           8     1.0000    0.9615    0.9804        52\n",
            "           9     0.9444    0.9444    0.9444        18\n",
            "          10     0.8889    1.0000    0.9412         8\n",
            "          11     1.0000    0.7895    0.8824        19\n",
            "          12     0.9968    0.9968    0.9968      1267\n",
            "          13     0.8889    0.6667    0.7619        12\n",
            "          14     1.0000    0.8784    0.9353        74\n",
            "          15     1.0000    1.0000    1.0000        27\n",
            "          16     1.0000    0.9574    0.9783        47\n",
            "\n",
            "    accuracy                         0.9967     13536\n",
            "   macro avg     0.9763    0.9188    0.9423     13536\n",
            "weighted avg     0.9967    0.9967    0.9966     13536\n",
            "\n",
            "KNN report for 7 Nearest Neighbours:\n",
            "              precision    recall  f1-score   support\n",
            "\n",
            "           0     0.9906    0.9968    0.9937       317\n",
            "           1     0.9512    0.9750    0.9630        40\n",
            "           2     0.9979    0.9983    0.9981      7024\n",
            "           3     0.9985    0.9987    0.9986      4573\n",
            "           4     1.0000    0.8182    0.9000        11\n",
            "           5     1.0000    0.9333    0.9655        15\n",
            "           6     1.0000    1.0000    1.0000        10\n",
            "           7     1.0000    1.0000    1.0000        22\n",
            "           8     0.9804    0.9615    0.9709        52\n",
            "           9     0.9000    1.0000    0.9474        18\n",
            "          10     1.0000    1.0000    1.0000         8\n",
            "          11     0.9500    1.0000    0.9744        19\n",
            "          12     0.9929    0.9953    0.9941      1267\n",
            "          13     1.0000    0.6667    0.8000        12\n",
            "          14     1.0000    0.9595    0.9793        74\n",
            "          15     1.0000    0.9259    0.9615        27\n",
            "          16     1.0000    1.0000    1.0000        47\n",
            "\n",
            "    accuracy                         0.9970     13536\n",
            "   macro avg     0.9860    0.9547    0.9674     13536\n",
            "weighted avg     0.9971    0.9970    0.9970     13536\n",
            "\n",
            "KNN report for 8 Nearest Neighbours:\n",
            "              precision    recall  f1-score   support\n",
            "\n",
            "           0     0.9968    0.9937    0.9953       317\n",
            "           1     1.0000    1.0000    1.0000        40\n",
            "           2     0.9982    0.9999    0.9990      7024\n",
            "           3     0.9983    0.9987    0.9985      4573\n",
            "           4     1.0000    0.8333    0.9091        12\n",
            "           5     1.0000    0.9333    0.9655        15\n",
            "           6     1.0000    0.9000    0.9474        10\n",
            "           7     0.9545    1.0000    0.9767        21\n",
            "           8     0.9804    0.9615    0.9709        52\n",
            "           9     1.0000    0.9444    0.9714        18\n",
            "          10     1.0000    1.0000    1.0000         7\n",
            "          11     1.0000    0.8947    0.9444        19\n",
            "          12     0.9961    1.0000    0.9980      1267\n",
            "          13     0.7778    0.5833    0.6667        12\n",
            "          14     1.0000    0.9595    0.9793        74\n",
            "          15     1.0000    0.9286    0.9630        28\n",
            "          16     1.0000    0.9362    0.9670        47\n",
            "\n",
            "    accuracy                         0.9977     13536\n",
            "   macro avg     0.9825    0.9334    0.9560     13536\n",
            "weighted avg     0.9977    0.9977    0.9976     13536\n",
            "\n",
            "KNN report for 9 Nearest Neighbours:\n",
            "              precision    recall  f1-score   support\n",
            "\n",
            "           0     0.9874    0.9905    0.9890       317\n",
            "           1     0.9744    0.9500    0.9620        40\n",
            "           2     0.9976    0.9990    0.9983      7024\n",
            "           3     0.9974    0.9983    0.9978      4573\n",
            "           4     1.0000    0.9167    0.9565        12\n",
            "           5     1.0000    0.9333    0.9655        15\n",
            "           6     0.9091    1.0000    0.9524        10\n",
            "           7     1.0000    1.0000    1.0000        21\n",
            "           8     0.9800    0.9423    0.9608        52\n",
            "           9     0.9444    0.9444    0.9444        18\n",
            "          10     1.0000    1.0000    1.0000         7\n",
            "          11     0.9375    0.7895    0.8571        19\n",
            "          12     0.9937    0.9976    0.9957      1267\n",
            "          13     0.8750    0.5833    0.7000        12\n",
            "          14     1.0000    0.9324    0.9650        74\n",
            "          15     1.0000    0.9286    0.9630        28\n",
            "          16     1.0000    0.9574    0.9783        47\n",
            "\n",
            "    accuracy                         0.9965     13536\n",
            "   macro avg     0.9763    0.9331    0.9521     13536\n",
            "weighted avg     0.9965    0.9965    0.9965     13536\n",
            "\n",
            "KNN report for 10 Nearest Neighbours:\n",
            "              precision    recall  f1-score   support\n",
            "\n",
            "           0     0.9873    0.9842    0.9858       317\n",
            "           1     0.9750    0.9750    0.9750        40\n",
            "           2     0.9980    0.9991    0.9986      7024\n",
            "           3     0.9983    0.9987    0.9985      4572\n",
            "           4     1.0000    0.8333    0.9091        12\n",
            "           5     0.8333    1.0000    0.9091        15\n",
            "           6     0.9000    0.9000    0.9000        10\n",
            "           7     1.0000    0.9545    0.9767        22\n",
            "           8     0.9623    0.9808    0.9714        52\n",
            "           9     1.0000    0.8889    0.9412        18\n",
            "          10     0.7778    1.0000    0.8750         7\n",
            "          11     1.0000    0.8947    0.9444        19\n",
            "          12     0.9976    0.9992    0.9984      1267\n",
            "          13     1.0000    0.7500    0.8571        12\n",
            "          14     1.0000    0.9730    0.9863        74\n",
            "          15     1.0000    0.9643    0.9818        28\n",
            "          16     1.0000    0.9149    0.9556        47\n",
            "\n",
            "    accuracy                         0.9972     13536\n",
            "   macro avg     0.9664    0.9418    0.9508     13536\n",
            "weighted avg     0.9973    0.9972    0.9972     13536\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xq08rdiIVaqo"
      },
      "source": [
        "In general, the algorithm actually perform best at k = 1. As the number of neighbour to take into consideration increases, the performance of the rare classes decreases, especially class 13. Recall for class 13 fluctuates wildly, meaning that the samples from this class are classified wrongly (high number of false negatives). "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tKRK6JRaYCxh",
        "outputId": "760ccf80-cf38-4dd1-a0ee-e6ae12faf8eb"
      },
      "source": [
        "#get actual phylum for class 13\n",
        "print(le.inverse_transform([13]))\n",
        "print(sum(y_encoded[y_encoded == 13]))"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[203682]\n",
            "2275\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bZC8cnWEYSFJ"
      },
      "source": [
        "The taxonomy ID belongs to Planctomycetes, with 2275 samples in our data. Planctomycetes have been known to occupy diverse niches, and are very metabolically flexible. Their flexibilty may lead to them occupying a dispersed area of the variable space, such that low number of nearest neighbours are able to correct assign the class, while when taking into account more neighbours encroach into the variable subspace occupied by other classes."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Xt-hvbnOXd8_",
        "outputId": "3863d0d6-ace3-49ce-b774-4e62af69890d"
      },
      "source": [
        "# knn on test set\n",
        "k = 1\n",
        "knn = KNeighborsClassifier(n_neighbors=k)\n",
        "knn.fit(X_train,y_train) #70% training and 30% test\n",
        "preds = knn.predict(X_test)\n",
        "knn_report = classification_report(y_test,preds,digits=4, output_dict=True)\n",
        "print(f\"KNN report for {k} Nearest Neighbours:\")\n",
        "print(classification_report(y_test, preds,digits=4))"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "KNN report for 1 Nearest Neighbours:\n",
            "              precision    recall  f1-score   support\n",
            "\n",
            "           0     0.9949    0.9963    0.9956      1359\n",
            "           1     0.9882    0.9882    0.9882       170\n",
            "           2     0.9990    0.9993    0.9992     30104\n",
            "           3     0.9990    0.9992    0.9991     19598\n",
            "           4     0.9800    0.9608    0.9703        51\n",
            "           5     1.0000    0.9524    0.9756        63\n",
            "           6     1.0000    0.9767    0.9882        43\n",
            "           7     0.9894    1.0000    0.9947        93\n",
            "           8     1.0000    0.9776    0.9887       223\n",
            "           9     1.0000    0.9870    0.9935        77\n",
            "          10     0.9697    1.0000    0.9846        32\n",
            "          11     0.9737    0.9136    0.9427        81\n",
            "          12     0.9982    0.9978    0.9980      5430\n",
            "          13     0.9434    0.9615    0.9524        52\n",
            "          14     0.9874    0.9874    0.9874       317\n",
            "          15     0.9831    0.9831    0.9831       118\n",
            "          16     1.0000    0.9852    0.9926       203\n",
            "\n",
            "    accuracy                         0.9986     58014\n",
            "   macro avg     0.9886    0.9804    0.9843     58014\n",
            "weighted avg     0.9986    0.9986    0.9985     58014\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "lMhwRYX6Ap_U"
      },
      "source": [
        "**Expected output:**\n",
        "\n",
        "```\n",
        "KNN report for 1 Nearest Neighbours:\n",
        "              precision    recall  f1-score   support\n",
        "\n",
        "           0     0.9949    0.9963    0.9956      1359\n",
        "           1     0.9882    0.9882    0.9882       170\n",
        "           2     0.9990    0.9993    0.9992     30104\n",
        "           3     0.9990    0.9992    0.9991     19598\n",
        "           4     0.9800    0.9608    0.9703        51\n",
        "           5     1.0000    0.9524    0.9756        63\n",
        "           6     1.0000    0.9767    0.9882        43\n",
        "           7     0.9894    1.0000    0.9947        93\n",
        "           8     1.0000    0.9776    0.9887       223\n",
        "           9     1.0000    0.9870    0.9935        77\n",
        "          10     0.9697    1.0000    0.9846        32\n",
        "          11     0.9737    0.9136    0.9427        81\n",
        "          12     0.9982    0.9978    0.9980      5430\n",
        "          13     0.9434    0.9615    0.9524        52\n",
        "          14     0.9874    0.9874    0.9874       317\n",
        "          15     0.9831    0.9831    0.9831       118\n",
        "          16     1.0000    0.9852    0.9926       203\n",
        "\n",
        "    accuracy                         0.9986     58014\n",
        "   macro avg     0.9886    0.9804    0.9843     58014\n",
        "weighted avg     0.9986    0.9986    0.9985     58014\n",
        "```"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "IE_MHbhfmLVT"
      },
      "source": [
        "### Training Random Forest classifier\n",
        "\n",
        "\n",
        "---\n",
        "\n",
        "Random forest is an ensemble learning method that operates on the consensus of multiple decision trees. In random forest method, there are many hyperparameters to tune for optimal performance. We will use `GridSearchCV` from `sklearn.model_selection` in order to select our best parameters. As hyperparameter tuning takes a lot of time especially with such a big data set, the blow code block is only a representative example of an optimising run. The full list of hyperparameters searched are as given below:\n",
        "\n",
        "```python\n",
        "param_grid = {\n",
        "    'bootstrap': [True],\n",
        "    'max_depth': [5,10,,20,25,30,35],\n",
        "    'max_features': [10,20,25,30,32,35,38,40,45],\n",
        "    'min_samples_leaf': [1,3,5,10],\n",
        "    'min_samples_split': [1,3,5,10],\n",
        "    'n_estimators': [25,50,75,100,125]\n",
        "}\n",
        "```"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Te4sRlqZdZqu",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "fbde90b2-8722-4512-f366-ac2dcc999107"
      },
      "source": [
        "from sklearn.ensemble import RandomForestClassifier\n",
        "from sklearn.model_selection import GridSearchCV\n",
        "\n",
        "rf = RandomForestClassifier(random_state = 42)\n",
        "\n",
        "X_train, X_test, y_train, y_test = train_test_split(X, y_encoded,stratify=y_encoded,test_size=0.3,random_state=42) # 70% training and 30% test\n",
        "\n",
        "param_grid = {\n",
        "    'bootstrap': [True],\n",
        "    'max_depth': [20,30],\n",
        "    'max_features': [35],\n",
        "    'min_samples_leaf': [5],\n",
        "    'min_samples_split': [5],\n",
        "    'n_estimators': [100]\n",
        "}\n",
        "\n",
        "# Instantiate the grid search model\n",
        "grid_search = GridSearchCV(estimator = rf, param_grid = param_grid, \n",
        "                          cv = 3, n_jobs = -1, verbose = 2)\n",
        "\n",
        "# Fit the grid search to the data\n",
        "grid_search.fit(X_train, y_train)\n",
        "grid_search.best_params_"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Fitting 3 folds for each of 2 candidates, totalling 6 fits\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.\n",
            "[Parallel(n_jobs=-1)]: Done   6 out of   6 | elapsed: 45.1min finished\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{'bootstrap': True,\n",
              " 'max_depth': 30,\n",
              " 'max_features': 35,\n",
              " 'min_samples_leaf': 5,\n",
              " 'min_samples_split': 5,\n",
              " 'n_estimators': 100}"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "UK7IydI21AbU"
      },
      "source": [
        "**Expected output**:\n",
        "```\n",
        "Fitting 3 folds for each of 2 candidates, totalling 6 fits\n",
        "[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.\n",
        "[Parallel(n_jobs=-1)]: Done   6 out of   6 | elapsed: 65.9min finished\n",
        "{'bootstrap': True,\n",
        " 'max_depth': 30,\n",
        " 'max_features': 35,\n",
        " 'min_samples_leaf': 5,\n",
        " 'min_samples_split': 5,\n",
        " 'n_estimators': 100}\n",
        " ```"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "UUjLq7_XnK7g",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "307deba1-6712-4bca-af35-8574b2a96494"
      },
      "source": [
        "# Using the best params from GridSearchCV\n",
        "rf = RandomForestClassifier(n_estimators=100, max_depth=30, max_features=35,\n",
        "                            min_samples_leaf=5,min_samples_split=5,random_state = 42)\n",
        "rf.fit(X_train,y_train)\n",
        "preds = rf.predict(X_test)\n",
        "rf_report = classification_report(y_test,preds,digits=4, output_dict=True)\n",
        "print(classification_report(y_test,preds,digits=4))\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "              precision    recall  f1-score   support\n",
            "\n",
            "           0     0.9926    0.9875    0.9900      1359\n",
            "           1     0.9880    0.9647    0.9762       170\n",
            "           2     0.9971    0.9992    0.9981     30104\n",
            "           3     0.9967    0.9985    0.9976     19598\n",
            "           4     1.0000    0.9608    0.9800        51\n",
            "           5     0.9828    0.9048    0.9421        63\n",
            "           6     1.0000    0.9535    0.9762        43\n",
            "           7     1.0000    1.0000    1.0000        93\n",
            "           8     0.9950    0.9013    0.9459       223\n",
            "           9     0.9737    0.9610    0.9673        77\n",
            "          10     1.0000    1.0000    1.0000        32\n",
            "          11     1.0000    0.7531    0.8592        81\n",
            "          12     0.9948    0.9954    0.9951      5430\n",
            "          13     0.9189    0.6538    0.7640        52\n",
            "          14     1.0000    0.9558    0.9774       317\n",
            "          15     0.9646    0.9237    0.9437       118\n",
            "          16     1.0000    0.9704    0.9850       203\n",
            "\n",
            "    accuracy                         0.9965     58014\n",
            "   macro avg     0.9885    0.9343    0.9587     58014\n",
            "weighted avg     0.9965    0.9965    0.9964     58014\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "izvKhNA-mLVT"
      },
      "source": [
        "### Training XGBoost classifier\n",
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "XGBoost is another ensemble method, where gradient boosting is incorporated to improve performance. This is done via gradient descent hence the name. Additionally, XGBoost gives the option of performing regularisation, preventing overfitting. We will utilise the hyperparameters fround from Random Forest and see if there is improvement with XGBoost. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "uBlmrH52s2Ak",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "95690cf9-0d2a-4f7a-d455-ff235095c64b"
      },
      "source": [
        "import xgboost as xgb\n",
        "from sklearn import preprocessing\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.metrics import f1_score\n",
        "\n",
        "# rename the classes\n",
        "le = preprocessing.LabelEncoder()\n",
        "y_encoded = le.fit_transform(y[:,0])\n",
        "\n",
        "X_train, X_test, y_train, y_test = train_test_split(X, y_encoded,stratify=y_encoded,test_size=0.2,random_state=42) # 80% training and 20% test\n",
        "X_train, X_eval, y_train, y_eval = train_test_split(X_train, y_train,stratify=y_train,test_size=0.25,random_state=42) # 60% training and 20% eval\n",
        "dtrain = xgb.DMatrix(X_train, label=y_train)\n",
        "deval = xgb.DMatrix(X_eval, label=y_eval)\n",
        "dtest = xgb.DMatrix(X_test, label=y_test)\n",
        "\n",
        "param = {\"objective\":\"multi:softmax\", \"n_estimators\":100, \"max_depth\":30, \n",
        "         \"min_child_weight\":5,\"lambda\":1.5, \"alpha\":1,\" colsample_bytree\":35/64, \n",
        "         \"learning_rate\":0.1, \"num_class\":17}\n",
        "evallist = [(dtrain, 'train'),(deval, 'eval')]\n",
        "xg_reg = xgb.train(param,dtrain,num_boost_round=1000,evals=evallist,early_stopping_rounds=10)"
      ],
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[0]\ttrain-merror:0.01073\teval-merror:0.014247\n",
            "Multiple eval metrics have been passed: 'eval-merror' will be used for early stopping.\n",
            "\n",
            "Will train until eval-merror hasn't improved in 10 rounds.\n",
            "[1]\ttrain-merror:0.007507\teval-merror:0.010006\n",
            "[2]\ttrain-merror:0.006283\teval-merror:0.008558\n",
            "[3]\ttrain-merror:0.005757\teval-merror:0.007886\n",
            "[4]\ttrain-merror:0.005232\teval-merror:0.007291\n",
            "[5]\ttrain-merror:0.004792\teval-merror:0.006878\n",
            "[6]\ttrain-merror:0.004318\teval-merror:0.006464\n",
            "[7]\ttrain-merror:0.004163\teval-merror:0.006283\n",
            "[8]\ttrain-merror:0.003887\teval-merror:0.006024\n",
            "[9]\ttrain-merror:0.003723\teval-merror:0.005895\n",
            "[10]\ttrain-merror:0.003577\teval-merror:0.005637\n",
            "[11]\ttrain-merror:0.003448\teval-merror:0.00543\n",
            "[12]\ttrain-merror:0.003284\teval-merror:0.005197\n",
            "[13]\ttrain-merror:0.003172\teval-merror:0.005016\n",
            "[14]\ttrain-merror:0.002999\teval-merror:0.004964\n",
            "[15]\ttrain-merror:0.002905\teval-merror:0.004861\n",
            "[16]\ttrain-merror:0.002784\teval-merror:0.004654\n",
            "[17]\ttrain-merror:0.002689\teval-merror:0.004525\n",
            "[18]\ttrain-merror:0.00262\teval-merror:0.004499\n",
            "[19]\ttrain-merror:0.002534\teval-merror:0.004499\n",
            "[20]\ttrain-merror:0.002413\teval-merror:0.004395\n",
            "[21]\ttrain-merror:0.002344\teval-merror:0.004421\n",
            "[22]\ttrain-merror:0.002224\teval-merror:0.004344\n",
            "[23]\ttrain-merror:0.002172\teval-merror:0.00437\n",
            "[24]\ttrain-merror:0.002094\teval-merror:0.00424\n",
            "[25]\ttrain-merror:0.002043\teval-merror:0.004215\n",
            "[26]\ttrain-merror:0.001974\teval-merror:0.004163\n",
            "[27]\ttrain-merror:0.001948\teval-merror:0.004111\n",
            "[28]\ttrain-merror:0.001888\teval-merror:0.004059\n",
            "[29]\ttrain-merror:0.001836\teval-merror:0.003956\n",
            "[30]\ttrain-merror:0.001698\teval-merror:0.003982\n",
            "[31]\ttrain-merror:0.001663\teval-merror:0.003723\n",
            "[32]\ttrain-merror:0.001612\teval-merror:0.003672\n",
            "[33]\ttrain-merror:0.001534\teval-merror:0.00362\n",
            "[34]\ttrain-merror:0.001465\teval-merror:0.003568\n",
            "[35]\ttrain-merror:0.001405\teval-merror:0.003516\n",
            "[36]\ttrain-merror:0.001353\teval-merror:0.003465\n",
            "[37]\ttrain-merror:0.001301\teval-merror:0.003439\n",
            "[38]\ttrain-merror:0.001267\teval-merror:0.003465\n",
            "[39]\ttrain-merror:0.00125\teval-merror:0.003413\n",
            "[40]\ttrain-merror:0.001164\teval-merror:0.003361\n",
            "[41]\ttrain-merror:0.001129\teval-merror:0.003361\n",
            "[42]\ttrain-merror:0.001086\teval-merror:0.00331\n",
            "[43]\ttrain-merror:0.001026\teval-merror:0.00331\n",
            "[44]\ttrain-merror:0.001008\teval-merror:0.003232\n",
            "[45]\ttrain-merror:0.000939\teval-merror:0.003206\n",
            "[46]\ttrain-merror:0.000888\teval-merror:0.003129\n",
            "[47]\ttrain-merror:0.000853\teval-merror:0.003077\n",
            "[48]\ttrain-merror:0.00081\teval-merror:0.003077\n",
            "[49]\ttrain-merror:0.000776\teval-merror:0.002948\n",
            "[50]\ttrain-merror:0.000767\teval-merror:0.002896\n",
            "[51]\ttrain-merror:0.000724\teval-merror:0.002792\n",
            "[52]\ttrain-merror:0.000698\teval-merror:0.002792\n",
            "[53]\ttrain-merror:0.000672\teval-merror:0.002767\n",
            "[54]\ttrain-merror:0.000646\teval-merror:0.002767\n",
            "[55]\ttrain-merror:0.000621\teval-merror:0.002689\n",
            "[56]\ttrain-merror:0.000621\teval-merror:0.002689\n",
            "[57]\ttrain-merror:0.000586\teval-merror:0.002611\n",
            "[58]\ttrain-merror:0.000577\teval-merror:0.002586\n",
            "[59]\ttrain-merror:0.000517\teval-merror:0.002586\n",
            "[60]\ttrain-merror:0.000491\teval-merror:0.00256\n",
            "[61]\ttrain-merror:0.000457\teval-merror:0.002508\n",
            "[62]\ttrain-merror:0.000431\teval-merror:0.002482\n",
            "[63]\ttrain-merror:0.000414\teval-merror:0.002482\n",
            "[64]\ttrain-merror:0.000405\teval-merror:0.002456\n",
            "[65]\ttrain-merror:0.000396\teval-merror:0.00243\n",
            "[66]\ttrain-merror:0.000371\teval-merror:0.002405\n",
            "[67]\ttrain-merror:0.000371\teval-merror:0.002353\n",
            "[68]\ttrain-merror:0.000328\teval-merror:0.002353\n",
            "[69]\ttrain-merror:0.000319\teval-merror:0.002353\n",
            "[70]\ttrain-merror:0.000319\teval-merror:0.002327\n",
            "[71]\ttrain-merror:0.000302\teval-merror:0.002327\n",
            "[72]\ttrain-merror:0.000276\teval-merror:0.002327\n",
            "[73]\ttrain-merror:0.000267\teval-merror:0.002249\n",
            "[74]\ttrain-merror:0.000267\teval-merror:0.002249\n",
            "[75]\ttrain-merror:0.000267\teval-merror:0.002249\n",
            "[76]\ttrain-merror:0.000267\teval-merror:0.002224\n",
            "[77]\ttrain-merror:0.000259\teval-merror:0.002224\n",
            "[78]\ttrain-merror:0.00025\teval-merror:0.002172\n",
            "[79]\ttrain-merror:0.000241\teval-merror:0.002172\n",
            "[80]\ttrain-merror:0.000233\teval-merror:0.002172\n",
            "[81]\ttrain-merror:0.000215\teval-merror:0.002146\n",
            "[82]\ttrain-merror:0.000215\teval-merror:0.002146\n",
            "[83]\ttrain-merror:0.000207\teval-merror:0.002146\n",
            "[84]\ttrain-merror:0.000207\teval-merror:0.002146\n",
            "[85]\ttrain-merror:0.000207\teval-merror:0.002146\n",
            "[86]\ttrain-merror:0.000207\teval-merror:0.002146\n",
            "[87]\ttrain-merror:0.000198\teval-merror:0.00212\n",
            "[88]\ttrain-merror:0.000198\teval-merror:0.00212\n",
            "[89]\ttrain-merror:0.00019\teval-merror:0.00212\n",
            "[90]\ttrain-merror:0.00019\teval-merror:0.00212\n",
            "[91]\ttrain-merror:0.00019\teval-merror:0.002146\n",
            "[92]\ttrain-merror:0.000172\teval-merror:0.002146\n",
            "[93]\ttrain-merror:0.000164\teval-merror:0.002146\n",
            "[94]\ttrain-merror:0.000164\teval-merror:0.002146\n",
            "[95]\ttrain-merror:0.000155\teval-merror:0.002094\n",
            "[96]\ttrain-merror:0.000155\teval-merror:0.002094\n",
            "[97]\ttrain-merror:0.000155\teval-merror:0.002094\n",
            "[98]\ttrain-merror:0.000155\teval-merror:0.002017\n",
            "[99]\ttrain-merror:0.000147\teval-merror:0.002043\n",
            "[100]\ttrain-merror:0.000147\teval-merror:0.002017\n",
            "[101]\ttrain-merror:0.000147\teval-merror:0.002017\n",
            "[102]\ttrain-merror:0.000138\teval-merror:0.002017\n",
            "[103]\ttrain-merror:0.000138\teval-merror:0.001939\n",
            "[104]\ttrain-merror:0.000129\teval-merror:0.001939\n",
            "[105]\ttrain-merror:0.000129\teval-merror:0.001939\n",
            "[106]\ttrain-merror:0.000129\teval-merror:0.001939\n",
            "[107]\ttrain-merror:0.000121\teval-merror:0.001939\n",
            "[108]\ttrain-merror:0.000121\teval-merror:0.001913\n",
            "[109]\ttrain-merror:0.000121\teval-merror:0.001939\n",
            "[110]\ttrain-merror:0.000121\teval-merror:0.001913\n",
            "[111]\ttrain-merror:0.000121\teval-merror:0.001913\n",
            "[112]\ttrain-merror:0.000121\teval-merror:0.001913\n",
            "[113]\ttrain-merror:0.000121\teval-merror:0.001887\n",
            "[114]\ttrain-merror:0.000121\teval-merror:0.001887\n",
            "[115]\ttrain-merror:0.000121\teval-merror:0.001862\n",
            "[116]\ttrain-merror:0.000121\teval-merror:0.001862\n",
            "[117]\ttrain-merror:0.000121\teval-merror:0.001862\n",
            "[118]\ttrain-merror:0.000121\teval-merror:0.001862\n",
            "[119]\ttrain-merror:0.000121\teval-merror:0.001836\n",
            "[120]\ttrain-merror:0.000112\teval-merror:0.001836\n",
            "[121]\ttrain-merror:0.000112\teval-merror:0.001862\n",
            "[122]\ttrain-merror:0.000112\teval-merror:0.001862\n",
            "[123]\ttrain-merror:0.000103\teval-merror:0.001836\n",
            "[124]\ttrain-merror:0.000103\teval-merror:0.001862\n",
            "[125]\ttrain-merror:0.000112\teval-merror:0.001862\n",
            "[126]\ttrain-merror:0.000103\teval-merror:0.00181\n",
            "[127]\ttrain-merror:0.000103\teval-merror:0.001862\n",
            "[128]\ttrain-merror:0.000103\teval-merror:0.001836\n",
            "[129]\ttrain-merror:0.000103\teval-merror:0.00181\n",
            "[130]\ttrain-merror:0.000103\teval-merror:0.00181\n",
            "[131]\ttrain-merror:0.000103\teval-merror:0.001758\n",
            "[132]\ttrain-merror:0.000103\teval-merror:0.001758\n",
            "[133]\ttrain-merror:0.000103\teval-merror:0.001784\n",
            "[134]\ttrain-merror:0.000103\teval-merror:0.001784\n",
            "[135]\ttrain-merror:0.000103\teval-merror:0.001784\n",
            "[136]\ttrain-merror:0.000103\teval-merror:0.001758\n",
            "[137]\ttrain-merror:0.000103\teval-merror:0.001732\n",
            "[138]\ttrain-merror:0.000103\teval-merror:0.001732\n",
            "[139]\ttrain-merror:0.000103\teval-merror:0.001732\n",
            "[140]\ttrain-merror:0.000103\teval-merror:0.001732\n",
            "[141]\ttrain-merror:0.000103\teval-merror:0.001732\n",
            "[142]\ttrain-merror:0.000103\teval-merror:0.001732\n",
            "[143]\ttrain-merror:0.000103\teval-merror:0.001732\n",
            "[144]\ttrain-merror:0.000103\teval-merror:0.001732\n",
            "[145]\ttrain-merror:0.000103\teval-merror:0.001732\n",
            "[146]\ttrain-merror:0.000103\teval-merror:0.001732\n",
            "[147]\ttrain-merror:0.000103\teval-merror:0.001732\n",
            "Stopping. Best iteration:\n",
            "[137]\ttrain-merror:0.000103\teval-merror:0.001732\n",
            "\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "error",
          "ename": "AttributeError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-23-e156cf41460c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0mevallist\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtrain\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'train'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdeval\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'eval'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0mxg_reg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mxgb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtrain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparam\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mdtrain\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mnum_boost_round\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1000\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mevals\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mevallist\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mearly_stopping_rounds\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 21\u001b[0;31m \u001b[0mpreds\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mxg_reg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpredict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_test\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     22\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[0;31m# evaluate predictions\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/xgboost/core.py\u001b[0m in \u001b[0;36mpredict\u001b[0;34m(self, data, output_margin, ntree_limit, pred_leaf, pred_contribs, approx_contribs, pred_interactions, validate_features)\u001b[0m\n\u001b[1;32m   1282\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1283\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mvalidate_features\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1284\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_features\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1285\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1286\u001b[0m         \u001b[0mlength\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mc_bst_ulong\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/xgboost/core.py\u001b[0m in \u001b[0;36m_validate_features\u001b[0;34m(self, data)\u001b[0m\n\u001b[1;32m   1673\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1674\u001b[0m             \u001b[0;31m# Booster can't accept data with different feature names\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1675\u001b[0;31m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfeature_names\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfeature_names\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1676\u001b[0m                 \u001b[0mdat_missing\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfeature_names\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfeature_names\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1677\u001b[0m                 \u001b[0mmy_missing\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfeature_names\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfeature_names\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mAttributeError\u001b[0m: 'numpy.ndarray' object has no attribute 'feature_names'"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "rgoFIJl-EGyC"
      },
      "source": [
        ""
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "SVWuEUKfEGML",
        "outputId": "8786217b-e7b1-4f4a-b160-2cfbe77f13a2"
      },
      "source": [
        "preds = xg_reg.predict(dtest)\n",
        "\n",
        "# evaluate predictions\n",
        "xgb_report = classification_report(y_test,preds,digits=4, output_dict=True)\n",
        "print(classification_report(y_test,preds,digits=4))"
      ],
      "execution_count": 30,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "              precision    recall  f1-score   support\n",
            "\n",
            "           0     0.9978    0.9948    0.9963      1359\n",
            "           1     1.0000    0.9706    0.9851       170\n",
            "           2     0.9987    0.9997    0.9992     30104\n",
            "           3     0.9988    0.9993    0.9990     19598\n",
            "           4     1.0000    1.0000    1.0000        51\n",
            "           5     0.9836    0.9524    0.9677        63\n",
            "           6     1.0000    0.9302    0.9639        43\n",
            "           7     1.0000    1.0000    1.0000        93\n",
            "           8     1.0000    0.9731    0.9864       223\n",
            "           9     1.0000    0.9870    0.9935        77\n",
            "          10     1.0000    1.0000    1.0000        32\n",
            "          11     1.0000    0.9630    0.9811        81\n",
            "          12     0.9989    0.9972    0.9981      5430\n",
            "          13     0.9583    0.8846    0.9200        52\n",
            "          14     0.9843    0.9905    0.9874       317\n",
            "          15     0.9915    0.9915    0.9915       118\n",
            "          16     1.0000    0.9803    0.9900       203\n",
            "\n",
            "    accuracy                         0.9986     58014\n",
            "   macro avg     0.9948    0.9773    0.9858     58014\n",
            "weighted avg     0.9986    0.9986    0.9986     58014\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    }
  ]
}